// @aztec/entrypoints/dest/entrypoint_payload.js
import { Fr } from '@aztec/aztec.js/fields';
import { PackedArguments, emptyFunctionCall } from '@aztec/circuit-types';
import { padArrayEnd } from '@aztec/foundation/collection';
import { pedersenHash } from '@aztec/foundation/crypto';
// These must match the values defined in:
// - noir-projects/aztec-nr/aztec/src/entrypoint/app.nr
const ACCOUNT_MAX_CALLS = 4;
// - and noir-projects/aztec-nr/aztec/src/entrypoint/fee.nr
const FEE_MAX_CALLS = 2;
/**
 * Builds a payload to be sent to the account contract
 * @param calls - The function calls to run
 * @param maxCalls - The maximum number of call expected to be run. Used for padding
 * @returns A payload object and packed arguments
 */
function buildPayload(calls, maxCalls) {
    const nonce = Fr.random();
    const paddedCalls = padArrayEnd(calls, emptyFunctionCall(), maxCalls);
    const packedArguments = [];
    for (const call of paddedCalls) {
        packedArguments.push(PackedArguments.fromArgs(call.args));
    }
    const formattedCalls = paddedCalls.map((call, index) => ({
        // eslint-disable-next-line camelcase
        args_hash: packedArguments[index].hash,
        // eslint-disable-next-line camelcase
        function_selector: call.functionData.selector.toField(),
        // eslint-disable-next-line camelcase
        target_address: call.to.toField(),
        // eslint-disable-next-line camelcase
        is_public: !call.functionData.isPrivate,
    }));
    return {
        payload: {
            // eslint-disable-next-line camelcase
            function_calls: formattedCalls,
            nonce,
        },
        packedArguments,
    };
}
/** builds the payload for a Dapp entrypoint */
export function buildDappPayload(call) {
    return buildPayload([call], 1);
}
/** Assembles an entrypoint app payload from a set of private and public function calls */
export function buildAppPayload(calls) {
    return buildPayload(calls, ACCOUNT_MAX_CALLS);
}
/** Creates the payload for paying the fee for a transaction */
export async function buildFeePayload(feeOpts) {
    const calls = feeOpts ? await feeOpts.paymentMethod.getFunctionCalls(new Fr(feeOpts.maxFee)) : [];
    return buildPayload(calls, FEE_MAX_CALLS);
}
// TODO (dogfooding) change all of these names app/dapp/fee/payload and generator indices for all of them
/** Hashes a payload to a 32-byte buffer */
export function hashPayload(payload, generatorIndex) {
    return pedersenHash(flattenPayload(payload).map(fr => fr.toBuffer()), generatorIndex);
}
/** Hash the payload for a dapp */
export function hashDappPayload(payload, userAddress, generatorIndex) {
    return pedersenHash([...flattenPayload(payload).map(fr => fr.toBuffer()), userAddress.toBuffer()], generatorIndex);
}
/** Flattens an payload */
function flattenPayload(payload) {
    return [
        ...payload.function_calls.flatMap(call => [
            call.args_hash,
            call.function_selector,
            call.target_address,
            new Fr(call.is_public),
        ]),
        payload.nonce,
    ];
}
