// @aztec/circuits.js/dest/structs/function_leaf_preimage.js
import { FunctionSelector } from '@aztec/foundation/abi';
import { pedersenHash } from '@aztec/foundation/crypto';
import { Fr } from '@aztec/foundation/fields';
import { BufferReader, serializeToBuffer, serializeToFields } from '@aztec/foundation/serialize';
import { FUNCTION_LEAF_PREIMAGE_LENGTH, GeneratorIndex } from '../constants.gen.js';
/**
 * A class representing the "preimage" of a function tree leaf.
 */
export class FunctionLeafPreimage {
    constructor(
    /**
     * Function selector.
     */
    functionSelector, 
    /**
     * Indicates whether the function is only callable by self or not.
     */
    isInternal, 
    /**
     * Indicates whether the function is private or public.
     */
    isPrivate, 
    /**
     * Verification key hash of the function.
     */
    vkHash, 
    /**
     * Hash of the ACIR of the function.
     */
    acirHash) {
        this.functionSelector = functionSelector;
        this.isInternal = isInternal;
        this.isPrivate = isPrivate;
        this.vkHash = vkHash;
        this.acirHash = acirHash;
    }
    static getFields(fields) {
        return [fields.functionSelector, fields.isInternal, fields.isPrivate, fields.vkHash, fields.acirHash];
    }
    /**
     * Serialize this as a buffer.
     * @returns The buffer.
     */
    toBuffer() {
        return serializeToBuffer(...FunctionLeafPreimage.getFields(this));
    }
    toFields() {
        const fields = serializeToFields(...FunctionLeafPreimage.getFields(this));
        if (fields.length !== FUNCTION_LEAF_PREIMAGE_LENGTH) {
            throw new Error(`Invalid number of fields for FunctionLeafPreimage. Expected ${FUNCTION_LEAF_PREIMAGE_LENGTH}, got ${fields.length}`);
        }
        return fields;
    }
    /**
     * Deserializes from a buffer or reader, corresponding to a write in cpp.
     * @param buffer - Buffer or reader to read from.
     * @returns A new instance of FunctionLeafPreimage.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new FunctionLeafPreimage(reader.readObject(FunctionSelector), reader.readBoolean(), reader.readBoolean(), Fr.fromBuffer(reader), Fr.fromBuffer(reader));
    }
    hash() {
        return pedersenHash(this.toFields().map(field => field.toBuffer()), GeneratorIndex.FUNCTION_LEAF);
    }
}
