// @aztec/circuits.js/dest/structs/global_variables.js
import { AztecAddress } from '@aztec/foundation/aztec-address';
import { EthAddress } from '@aztec/foundation/eth-address';
import { Fr } from '@aztec/foundation/fields';
import { BufferReader, FieldReader, serializeToBuffer, serializeToFields } from '@aztec/foundation/serialize';
import { GLOBAL_VARIABLES_LENGTH } from '../constants.gen.js';
/**
 * Global variables of the L2 block.
 */
export class GlobalVariables {
    constructor(
    /** ChainId for the L2 block. */
    chainId, 
    /** Version for the L2 block. */
    version, 
    /** Block number of the L2 block. */
    blockNumber, 
    /** Timestamp of the L2 block. */
    timestamp, 
    /** Recipient of block reward. */
    coinbase, 
    /** Address to receive fees. */
    feeRecipient) {
        this.chainId = chainId;
        this.version = version;
        this.blockNumber = blockNumber;
        this.timestamp = timestamp;
        this.coinbase = coinbase;
        this.feeRecipient = feeRecipient;
    }
    static from(fields) {
        return new GlobalVariables(...GlobalVariables.getFields(fields));
    }
    static empty() {
        return new GlobalVariables(Fr.ZERO, Fr.ZERO, Fr.ZERO, Fr.ZERO, EthAddress.ZERO, AztecAddress.ZERO);
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new GlobalVariables(Fr.fromBuffer(reader), Fr.fromBuffer(reader), Fr.fromBuffer(reader), Fr.fromBuffer(reader), reader.readObject(EthAddress), reader.readObject(AztecAddress));
    }
    static fromJSON(obj) {
        return new GlobalVariables(Fr.fromString(obj.chainId), Fr.fromString(obj.version), Fr.fromString(obj.blockNumber), Fr.fromString(obj.timestamp), EthAddress.fromString(obj.coinbase), AztecAddress.fromString(obj.feeRecipient));
    }
    static fromFields(fields) {
        const reader = FieldReader.asReader(fields);
        return new GlobalVariables(reader.readField(), reader.readField(), reader.readField(), reader.readField(), EthAddress.fromField(reader.readField()), AztecAddress.fromField(reader.readField()));
    }
    static getFields(fields) {
        // Note: The order here must match the order in the HeaderLib solidity library.
        return [
            fields.chainId,
            fields.version,
            fields.blockNumber,
            fields.timestamp,
            fields.coinbase,
            fields.feeRecipient,
        ];
    }
    toBuffer() {
        return serializeToBuffer(...GlobalVariables.getFields(this));
    }
    toFields() {
        const fields = serializeToFields(...GlobalVariables.getFields(this));
        if (fields.length !== GLOBAL_VARIABLES_LENGTH) {
            throw new Error(`Invalid number of fields for GlobalVariables. Expected ${GLOBAL_VARIABLES_LENGTH}, got ${fields.length}`);
        }
        return fields;
    }
    toJSON() {
        return {
            chainId: this.chainId.toString(),
            version: this.version.toString(),
            blockNumber: this.blockNumber.toString(),
            timestamp: this.timestamp.toString(),
            coinbase: this.coinbase.toString(),
            feeRecipient: this.feeRecipient.toString(),
        };
    }
    isEmpty() {
        return (this.chainId.isZero() &&
            this.version.isZero() &&
            this.blockNumber.isZero() &&
            this.timestamp.isZero() &&
            this.coinbase.isZero() &&
            this.feeRecipient.isZero());
    }
}
