// @aztec/circuits.js/dest/structs/header.js
import { pedersenHash } from '@aztec/foundation/crypto';
import { BufferReader, FieldReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { GeneratorIndex, HEADER_LENGTH } from '../constants.gen.js';
import { ContentCommitment } from './content_commitment.js';
import { GlobalVariables } from './global_variables.js';
import { AppendOnlyTreeSnapshot } from './rollup/append_only_tree_snapshot.js';
import { StateReference } from './state_reference.js';
/** A header of an L2 block. */
export class Header {
    constructor(
    /** Snapshot of archive before the block is applied. */
    lastArchive, 
    /** Hash of the body of an L2 block. */
    contentCommitment, 
    /** State reference. */
    state, 
    /** Global variables of an L2 block. */
    globalVariables) {
        this.lastArchive = lastArchive;
        this.contentCommitment = contentCommitment;
        this.state = state;
        this.globalVariables = globalVariables;
    }
    toBuffer() {
        // Note: The order here must match the order in the HeaderLib solidity library.
        return serializeToBuffer(this.lastArchive, this.contentCommitment, this.state, this.globalVariables);
    }
    toFields() {
        // Note: The order here must match the order in header.nr
        const fields = [
            ...this.lastArchive.toFields(),
            ...this.contentCommitment.toFields(),
            ...this.state.toFields(),
            ...this.globalVariables.toFields(),
        ];
        if (fields.length !== HEADER_LENGTH) {
            throw new Error(`Invalid number of fields for Header. Expected ${HEADER_LENGTH}, got ${fields.length}`);
        }
        return fields;
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new Header(reader.readObject(AppendOnlyTreeSnapshot), reader.readObject(ContentCommitment), reader.readObject(StateReference), reader.readObject(GlobalVariables));
    }
    static fromFields(fields) {
        const reader = FieldReader.asReader(fields);
        const lastArchive = new AppendOnlyTreeSnapshot(reader.readField(), Number(reader.readField().toBigInt()));
        const contentCommitment = ContentCommitment.fromFields(reader);
        const state = StateReference.fromFields(reader);
        const globalVariables = GlobalVariables.fromFields(reader);
        return new Header(lastArchive, contentCommitment, state, globalVariables);
    }
    static empty() {
        return new Header(AppendOnlyTreeSnapshot.zero(), ContentCommitment.empty(), StateReference.empty(), GlobalVariables.empty());
    }
    isEmpty() {
        return (this.lastArchive.isZero() &&
            this.contentCommitment.isEmpty() &&
            this.state.isEmpty() &&
            this.globalVariables.isEmpty());
    }
    /**
     * Serializes this instance into a string.
     * @returns Encoded string.
     */
    toString() {
        return this.toBuffer().toString('hex');
    }
    static fromString(str) {
        const buffer = Buffer.from(str.replace(/^0x/i, ''), 'hex');
        return Header.fromBuffer(buffer);
    }
    hash() {
        return pedersenHash(this.toFields().map(f => f.toBuffer()), GeneratorIndex.BLOCK_HASH);
    }
}
