// @aztec/circuits.js/dest/structs/private_call_stack_item.js
import { AztecAddress } from '@aztec/foundation/aztec-address';
import { pedersenHash } from '@aztec/foundation/crypto';
import { Fr } from '@aztec/foundation/fields';
import { BufferReader, FieldReader, serializeToBuffer, serializeToFields } from '@aztec/foundation/serialize';
import { GeneratorIndex, PRIVATE_CALL_STACK_ITEM_LENGTH } from '../constants.gen.js';
import { CallRequest, CallerContext } from './call_request.js';
import { FunctionData } from './function_data.js';
import { PrivateCircuitPublicInputs } from './private_circuit_public_inputs.js';
/**
 * Call stack item on a private call.
 */
export class PrivateCallStackItem {
    constructor(
    /**
     * Address of the contract on which the function is invoked.
     */
    contractAddress, 
    /**
     * Data identifying the function being called.
     */
    functionData, 
    /**
     * Public inputs to the private kernel circuit.
     */
    publicInputs) {
        this.contractAddress = contractAddress;
        this.functionData = functionData;
        this.publicInputs = publicInputs;
    }
    static getFields(fields) {
        return [fields.contractAddress, fields.functionData, fields.publicInputs];
    }
    toBuffer() {
        return serializeToBuffer(...PrivateCallStackItem.getFields(this));
    }
    toFields() {
        const fields = serializeToFields(...PrivateCallStackItem.getFields(this));
        if (fields.length !== PRIVATE_CALL_STACK_ITEM_LENGTH) {
            throw new Error(`Invalid number of fields for PrivateCallStackItem. Expected ${PRIVATE_CALL_STACK_ITEM_LENGTH}, got ${fields.length}`);
        }
        return fields;
    }
    /**
     * Deserializes from a buffer or reader.
     * @param buffer - Buffer or reader to read from.
     * @returns The deserialized instance.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new PrivateCallStackItem(reader.readObject(AztecAddress), reader.readObject(FunctionData), reader.readObject(PrivateCircuitPublicInputs));
    }
    static fromFields(fields) {
        const reader = FieldReader.asReader(fields);
        return new PrivateCallStackItem(AztecAddress.fromFields(reader), FunctionData.fromFields(reader), PrivateCircuitPublicInputs.fromFields(reader));
    }
    /**
     * Returns a new instance of PrivateCallStackItem with zero contract address, function data and public inputs.
     * @returns A new instance of PrivateCallStackItem with zero contract address, function data and public inputs.
     */
    static empty() {
        return new PrivateCallStackItem(AztecAddress.ZERO, FunctionData.empty({ isPrivate: true }), PrivateCircuitPublicInputs.empty());
    }
    isEmpty() {
        return this.contractAddress.isZero() && this.functionData.isEmpty() && this.publicInputs.isEmpty();
    }
    /**
     * Computes this call stack item hash.
     * @returns Hash.
     */
    hash() {
        return pedersenHash(this.toFields().map(field => field.toBuffer()), GeneratorIndex.CALL_STACK_ITEM);
    }
    /**
     * Creates a new CallRequest with values of the calling contract.
     * @returns A CallRequest instance with the contract address, caller context, and the hash of the call stack item.
     */
    toCallRequest(parentCallContext) {
        if (this.isEmpty()) {
            return CallRequest.empty();
        }
        const currentCallContext = this.publicInputs.callContext;
        const callerContext = currentCallContext.isDelegateCall
            ? new CallerContext(parentCallContext.msgSender, parentCallContext.storageContractAddress)
            : CallerContext.empty();
        return new CallRequest(this.hash(), parentCallContext.storageContractAddress, callerContext, new Fr(this.publicInputs.callContext.sideEffectCounter), this.publicInputs.endSideEffectCounter);
    }
}
