// @aztec/circuits.js/dest/structs/public_data_read_request.js
import { Fr } from '@aztec/foundation/fields';
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
// TODO: Rename to PublicDataReadRequest
/**
 * Read operations from the public state tree.
 */
export class PublicDataRead {
    constructor(
    /**
     * Index of the leaf in the public data tree.
     */
    leafSlot, 
    /**
     * Returned value from the public data tree.
     */
    value, 
    /**
     * Optional side effect counter tracking position of this event in tx execution.
     */
    sideEffectCounter) {
        this.leafSlot = leafSlot;
        this.value = value;
        this.sideEffectCounter = sideEffectCounter;
    }
    static from(args) {
        return new PublicDataRead(args.leafIndex, args.value);
    }
    toBuffer() {
        return serializeToBuffer(this.leafSlot, this.value);
    }
    isEmpty() {
        return this.leafSlot.isZero() && this.value.isZero();
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new PublicDataRead(Fr.fromBuffer(reader), Fr.fromBuffer(reader));
    }
    static empty() {
        return new PublicDataRead(Fr.ZERO, Fr.ZERO);
    }
    toFriendlyJSON() {
        return `Leaf=${this.leafSlot.toFriendlyJSON()}: ${this.value.toFriendlyJSON()}`;
    }
    equals(other) {
        return this.leafSlot.equals(other.leafSlot) && this.value.equals(other.value);
    }
}
