// @aztec/circuits.js/dest/structs/public_call_stack_item.js
import { AztecAddress } from '@aztec/foundation/aztec-address';
import { pedersenHash } from '@aztec/foundation/crypto';
import { Fr } from '@aztec/foundation/fields';
import { BufferReader, FieldReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { GeneratorIndex } from '../constants.gen.js';
import { CallRequest, CallerContext } from './call_request.js';
import { FunctionData } from './function_data.js';
import { PublicCircuitPublicInputs } from './public_circuit_public_inputs.js';
/**
 * Call stack item on a public call.
 */
export class PublicCallStackItem {
    constructor(
    /**
     * Address of the contract on which the function is invoked.
     */
    contractAddress, 
    /**
     * Data identifying the function being called.
     */
    functionData, 
    /**
     * Public inputs to the public kernel circuit.
     */
    publicInputs, 
    /**
     * Whether the current callstack item should be considered a public fn execution request.
     */
    isExecutionRequest) {
        this.contractAddress = contractAddress;
        this.functionData = functionData;
        this.publicInputs = publicInputs;
        this.isExecutionRequest = isExecutionRequest;
    }
    static getFields(fields) {
        return [fields.contractAddress, fields.functionData, fields.publicInputs, fields.isExecutionRequest];
    }
    toBuffer() {
        return serializeToBuffer(...PublicCallStackItem.getFields(this));
    }
    /**
     * Deserializes from a buffer or reader.
     * @param buffer - Buffer or reader to read from.
     * @returns The deserialized instance.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new PublicCallStackItem(reader.readObject(AztecAddress), reader.readObject(FunctionData), reader.readObject(PublicCircuitPublicInputs), reader.readBoolean());
    }
    static fromFields(fields) {
        const reader = FieldReader.asReader(fields);
        const contractAddress = AztecAddress.fromFields(reader);
        const functionData = FunctionData.fromFields(reader);
        const publicInputs = PublicCircuitPublicInputs.fromFields(reader);
        const isExecutionRequest = reader.readBoolean();
        return new PublicCallStackItem(contractAddress, functionData, publicInputs, isExecutionRequest);
    }
    /**
     * Returns a new instance of PublicCallStackItem with zero contract address, function data and public inputs.
     * @returns A new instance of PublicCallStackItem with zero contract address, function data and public inputs.
     */
    static empty() {
        return new PublicCallStackItem(AztecAddress.ZERO, FunctionData.empty({ isPrivate: false }), PublicCircuitPublicInputs.empty(), false);
    }
    isEmpty() {
        return this.contractAddress.isZero() && this.functionData.isEmpty() && this.publicInputs.isEmpty();
    }
    /**
     * Computes this call stack item hash.
     * @returns Hash.
     */
    hash() {
        if (this.isExecutionRequest) {
            const { callContext, argsHash } = this.publicInputs;
            this.publicInputs = PublicCircuitPublicInputs.empty();
            this.publicInputs.callContext = callContext;
            this.publicInputs.argsHash = argsHash;
        }
        return pedersenHash([this.contractAddress, this.functionData.hash(), this.publicInputs.hash()].map(f => f.toBuffer()), GeneratorIndex.CALL_STACK_ITEM);
    }
    /**
     * Creates a new CallRequest with values of the calling contract.
     * @returns A CallRequest instance with the contract address, caller context, and the hash of the call stack item.
     */
    toCallRequest(parentCallContext) {
        if (this.isEmpty()) {
            return CallRequest.empty();
        }
        const currentCallContext = this.publicInputs.callContext;
        const callerContext = currentCallContext.isDelegateCall
            ? new CallerContext(parentCallContext.msgSender, parentCallContext.storageContractAddress)
            : CallerContext.empty();
        // todo: populate side effect counters correctly
        return new CallRequest(this.hash(), parentCallContext.storageContractAddress, callerContext, Fr.ZERO, Fr.ZERO);
    }
}
