// @aztec/circuits.js/dest/structs/side_effects.js
import { Fr } from '@aztec/foundation/fields';
import { BufferReader, FieldReader, serializeToBuffer } from '@aztec/foundation/serialize';
/**
 * Side-effect object consisting of a value and a counter.
 * cpp/src/aztec3/circuits/abis/side_effects.hpp.
 */
export class SideEffect {
    constructor(
    /**
     * The value of the side-effect object.
     */
    value, 
    /**
     * The side-effect counter.
     */
    counter) {
        this.value = value;
        this.counter = counter;
    }
    toString() {
        return `value=${this.value.toString()} counter=${this.counter.toString()}`;
    }
    /**
     * Serialize this as a buffer.
     * @returns The buffer.
     */
    toBuffer() {
        return serializeToBuffer(this.value, this.counter);
    }
    /**
     * Convert to an array of fields.
     * @returns The array of fields.
     */
    toFields() {
        return [this.value, this.counter];
    }
    static fromFields(fields) {
        const reader = FieldReader.asReader(fields);
        return new SideEffect(reader.readField(), reader.readField());
    }
    /**
     * Returns whether this instance of side-effect is empty.
     * @returns True if the value and counter both are zero.
     */
    isEmpty() {
        return SideEffect.isEmpty(this);
    }
    /**
     * Checks whether this instance of side-effect is empty.
     * @returns True if the value and counter both are zero.
     */
    static isEmpty(sideEffect) {
        return sideEffect.value.isZero() && sideEffect.counter.isZero();
    }
    /**
     * Returns an empty instance of side-effect.
     * @returns Side-effect with both value and counter being zero.
     */
    static empty() {
        return new SideEffect(Fr.zero(), Fr.zero());
    }
    /**
     * Deserializes from a buffer or reader, corresponding to a write in cpp.
     * @param buffer - Buffer or reader to read from.
     * @returns A new instance of SideEffect.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new SideEffect(Fr.fromBuffer(reader), Fr.fromBuffer(reader));
    }
}
/**
 * Side-effect object consisting of a value, a start counter and an end counter.
 * cpp/src/aztec3/circuits/abis/side_effects.hpp.
 */
export class SideEffectLinkedToNoteHash {
    constructor(
    /**
     * The value of the side-effect object.
     */
    value, 
    /**
     * The note hash corresponding to the side-effect value.
     */
    noteHash, 
    /**
     * The counter.
     */
    counter) {
        this.value = value;
        this.noteHash = noteHash;
        this.counter = counter;
    }
    toString() {
        return `value=${this.value.toString()} noteHash=${this.noteHash.toString()} counter=${this.counter.toString()}`;
    }
    /**
     * Serialize this as a buffer.
     * @returns The buffer.
     */
    toBuffer() {
        return serializeToBuffer(this.value, this.noteHash, this.counter);
    }
    /**
     * Convert to an array of fields.
     * @returns The array of fields.
     */
    toFields() {
        return [this.value, this.noteHash, this.counter];
    }
    static fromFields(fields) {
        const reader = FieldReader.asReader(fields);
        return new SideEffectLinkedToNoteHash(reader.readField(), reader.readField(), reader.readField());
    }
    /**
     * Returns whether this instance of side-effect is empty.
     * @returns True if the value, note hash and counter are all zero.
     */
    isEmpty() {
        return SideEffectLinkedToNoteHash.isEmpty(this);
    }
    /**
     * Returns whether this instance of side-effect is empty.
     * @returns True if the value, note hash and counter are all zero.
     */
    static isEmpty(sideEffect) {
        return sideEffect.value.isZero() && sideEffect.noteHash.isZero() && sideEffect.counter.isZero();
    }
    /**
     * Returns an empty instance of side-effect.
     * @returns Side-effect with value, note hash and counter being zero.
     */
    static empty() {
        return new SideEffectLinkedToNoteHash(Fr.zero(), Fr.zero(), Fr.zero());
    }
    /**
     * Deserializes from a buffer or reader, corresponding to a write in cpp.
     * @param buffer - Buffer or reader to read from.
     * @returns A new instance of SideEffectLinkedToNoteHash.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new SideEffectLinkedToNoteHash(Fr.fromBuffer(reader), Fr.fromBuffer(reader), Fr.fromBuffer(reader));
    }
}
/**
 * Convert an array of side effects to an array only non-empty side effects.
 * @param sideEffects - array to be converted
 * @returns the array of the non-empty side effects
 */
export function nonEmptySideEffects(sideEffects) {
    return sideEffects.filter(sideEffect => !sideEffect.isEmpty());
}
/**
 * Convert an array of side effects to an array of their values.
 * @param sideEffects - array to be converted
 * @returns the array of field values (excluding SideEffect metadata like counter)
 */
export function sideEffectArrayToValueArray(sideEffects) {
    return sideEffects.map(sideEffect => sideEffect.value);
}
/**
 * Compare two side effects based on their counter.
 * @param a - A side effect
 * @param b - Another side effect
 * @returns - The order of the two side effects
 */
export function sideEffectCmp(a, b) {
    return a.counter.cmp(b.counter);
}
