// @aztec/circuits.js/dest/structs/tx_context.js
import { pedersenHash } from '@aztec/foundation/crypto';
import { Fr } from '@aztec/foundation/fields';
import { BufferReader, serializeToBuffer, serializeToFields } from '@aztec/foundation/serialize';
import { GeneratorIndex, TX_CONTEXT_DATA_LENGTH } from '../constants.gen.js';
/**
 * Transaction context.
 */
export class TxContext {
    constructor(
    /**
     * Whether this is a fee paying tx. If not other tx in a bundle will pay the fee.
     * TODO(#3417): Remove fee and rebate payment fields.
     */
    isFeePaymentTx, 
    /**
     * Indicates whether this a gas rebate payment tx.
     *
     * NOTE: The following is a WIP and it is likely to change in the future.
     * Explanation: Each tx is actually 3 txs in one: a fee-paying tx, the actual tx you want to execute, and a rebate
     * tx. The fee-paying tx pays some `max_fee = gas_price * gas_limit`. Then the actual tx will cost an amount of gas
     * to execute (actual_fee = gas_price * gas_used). Then the rebate tx returns `max_fee - actual_fee` back to
     * the user.
     */
    isRebatePaymentTx, 
    /**
     * Chain ID of the transaction. Here for replay protection.
     */
    chainId, 
    /**
     * Version of the transaction. Here for replay protection.
     */
    version) {
        this.isFeePaymentTx = isFeePaymentTx;
        this.isRebatePaymentTx = isRebatePaymentTx;
        this.chainId = chainId;
        this.version = version;
    }
    /**
     * Serialize as a buffer.
     * @returns The buffer.
     */
    toBuffer() {
        return serializeToBuffer(...TxContext.getFields(this));
    }
    toFields() {
        const fields = serializeToFields(...TxContext.getFields(this));
        if (fields.length !== TX_CONTEXT_DATA_LENGTH) {
            throw new Error(`Invalid number of fields for TxContext. Expected ${TX_CONTEXT_DATA_LENGTH}, got ${fields.length}`);
        }
        return fields;
    }
    /**
     * Deserializes TxContext from a buffer or reader.
     * @param buffer - Buffer to read from.
     * @returns The TxContext.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new TxContext(reader.readBoolean(), reader.readBoolean(), Fr.fromBuffer(reader), Fr.fromBuffer(reader));
    }
    static empty(chainId = 0, version = 0) {
        return new TxContext(false, false, new Fr(chainId), new Fr(version));
    }
    isEmpty() {
        return !this.isFeePaymentTx && !this.isRebatePaymentTx && this.chainId.isZero() && this.version.isZero();
    }
    /**
     * Create a new instance from a fields dictionary.
     * @param fields - The dictionary.
     * @returns A new instance.
     */
    static from(fields) {
        return new TxContext(...TxContext.getFields(fields));
    }
    /**
     * Serialize into a field array. Low-level utility.
     * @param fields - Object with fields.
     * @returns The array.
     */
    static getFields(fields) {
        return [fields.isFeePaymentTx, fields.isRebatePaymentTx, fields.chainId, fields.version];
    }
    hash() {
        return pedersenHash(this.toFields().map(f => f.toBuffer()), GeneratorIndex.TX_CONTEXT);
    }
}
