// @aztec/circuits.js/dest/structs/parity/parity_public_inputs.js
import { Fr } from '@aztec/foundation/fields';
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { AggregationObject } from '../aggregation_object.js';
export class ParityPublicInputs {
    constructor(
    /** Aggregated proof of all the parity circuit iterations. */
    aggregationObject, 
    /** Root of the SHA256 tree. */
    shaRoot, 
    /** Root of the converted tree. */
    convertedRoot) {
        this.aggregationObject = aggregationObject;
        this.shaRoot = shaRoot;
        this.convertedRoot = convertedRoot;
        if (shaRoot.length !== 32) {
            throw new Error(`shaRoot buffer must be 32 bytes. Got ${shaRoot.length} bytes`);
        }
    }
    toBuffer() {
        return serializeToBuffer(...ParityPublicInputs.getFields(this));
    }
    static from(fields) {
        return new ParityPublicInputs(...ParityPublicInputs.getFields(fields));
    }
    static getFields(fields) {
        return [fields.aggregationObject, fields.shaRoot, fields.convertedRoot];
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new ParityPublicInputs(reader.readObject(AggregationObject), reader.readBytes(32), reader.readObject(Fr));
    }
}
