// @aztec/circuits.js/dest/contract/contract_class.js
import { FunctionSelector, FunctionType } from '@aztec/foundation/abi';
import { Fr } from '@aztec/foundation/fields';
import { computeArtifactHash } from './artifact_hash.js';
import { computeContractClassIdWithPreimage } from './contract_class_id.js';
import { packBytecode } from './public_bytecode.js';
const cmpFunctionArtifacts = (a, b) => a.selector.toField().cmp(b.selector.toField());
/** Creates a ContractClass from a contract compilation artifact. */
export function getContractClassFromArtifact(artifact) {
    const artifactHash = 'artifactHash' in artifact ? artifact.artifactHash : computeArtifactHash(artifact);
    const publicFunctions = artifact.functions
        .filter(f => f.functionType === FunctionType.OPEN)
        .map(f => ({
        selector: FunctionSelector.fromNameAndParameters(f.name, f.parameters),
        bytecode: Buffer.from(f.bytecode, 'base64'),
        isInternal: f.isInternal,
    }))
        .sort(cmpFunctionArtifacts);
    const packedBytecode = packBytecode(publicFunctions);
    const privateFunctions = artifact.functions
        .filter(f => f.functionType === FunctionType.SECRET)
        .map(f => ({
        selector: FunctionSelector.fromNameAndParameters(f.name, f.parameters),
        vkHash: getVerificationKeyHash(f.verificationKey),
        isInternal: f.isInternal,
    }))
        .sort(cmpFunctionArtifacts);
    const contractClass = {
        version: 1,
        artifactHash,
        publicFunctions,
        packedBytecode,
        privateFunctions,
    };
    return { ...contractClass, ...computeContractClassIdWithPreimage(contractClass) };
}
/**
 * Calculates the hash of a verification key.
 * Returns zero for consistency with Noir.
 */
function getVerificationKeyHash(_verificationKeyInBase64) {
    // return Fr.fromBuffer(hashVK(Buffer.from(verificationKeyInBase64, 'hex')));
    return Fr.ZERO;
}
