// @aztec/circuits.js/dest/structs/kernel/private_kernel_inner_circuit_public_inputs.js
import { Fr } from '@aztec/foundation/fields';
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { AggregationObject } from '../aggregation_object.js';
import { ValidationRequests } from '../validation_requests.js';
import { CombinedAccumulatedData } from './combined_accumulated_data.js';
import { CombinedConstantData } from './combined_constant_data.js';
/**
 * Public inputs to the inner private kernel circuit
 */
export class PrivateKernelInnerCircuitPublicInputs {
    constructor(
    /**
     * Aggregated proof of all the previous kernel iterations.
     */
    aggregationObject, // Contains the aggregated proof of all previous kernel iterations
    /**
     * The side effect counter that non-revertible side effects are all beneath.
     */
    minRevertibleSideEffectCounter, 
    /**
     * Validation requests accumulated from public functions.
     */
    validationRequests, 
    /**
     * Data accumulated from both public and private circuits.
     */
    end, 
    /**
     * Data which is not modified by the circuits.
     */
    constants, 
    /**
     * Indicates whether the input is for a private or public kernel.
     */
    isPrivate) {
        this.aggregationObject = aggregationObject;
        this.minRevertibleSideEffectCounter = minRevertibleSideEffectCounter;
        this.validationRequests = validationRequests;
        this.end = end;
        this.constants = constants;
        this.isPrivate = isPrivate;
    }
    toBuffer() {
        return serializeToBuffer(this.aggregationObject, this.minRevertibleSideEffectCounter, this.validationRequests, this.end, this.constants, this.isPrivate);
    }
    /**
     * Deserializes from a buffer or reader, corresponding to a write in cpp.
     * @param buffer - Buffer or reader to read from.
     * @returns A new instance of PrivateKernelInnerCircuitPublicInputs.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new PrivateKernelInnerCircuitPublicInputs(reader.readObject(AggregationObject), reader.readObject(Fr), reader.readObject(ValidationRequests), reader.readObject(CombinedAccumulatedData), reader.readObject(CombinedConstantData), reader.readBoolean());
    }
    static empty() {
        return new PrivateKernelInnerCircuitPublicInputs(AggregationObject.makeFake(), Fr.zero(), ValidationRequests.empty(), CombinedAccumulatedData.empty(), CombinedConstantData.empty(), true);
    }
}
