// @aztec/circuits.js/dest/structs/kernel/private_kernel_tail_circuit_public_inputs.js
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { AggregationObject } from '../aggregation_object.js';
import { PrivateAccumulatedNonRevertibleData, PrivateAccumulatedRevertibleData } from './combined_accumulated_data.js';
import { CombinedConstantData } from './combined_constant_data.js';
/**
 * Output from to the private kernel circuit - tail call.
 */
export class PrivateKernelTailCircuitPublicInputs {
    constructor(
    /**
     * Aggregated proof of all the previous kernel iterations.
     */
    aggregationObject, // Contains the aggregated proof of all previous kernel iterations
    /**
     * Accumulated side effects that are not revertible.
     */
    endNonRevertibleData, 
    /**
     * Data accumulated from both public and private circuits.
     */
    end, 
    /**
     * Data which is not modified by the circuits.
     */
    constants, 
    /**
     * Indicates whether the setup kernel is needed.
     */
    needsSetup, 
    /**
     * Indicates whether the app logic kernel is needed.
     */
    needsAppLogic, 
    /**
     * Indicates whether the teardown kernel is needed.
     */
    needsTeardown) {
        this.aggregationObject = aggregationObject;
        this.endNonRevertibleData = endNonRevertibleData;
        this.end = end;
        this.constants = constants;
        this.needsSetup = needsSetup;
        this.needsAppLogic = needsAppLogic;
        this.needsTeardown = needsTeardown;
    }
    toBuffer() {
        return serializeToBuffer(this.aggregationObject, this.endNonRevertibleData, this.end, this.constants, this.needsSetup, this.needsAppLogic, this.needsTeardown);
    }
    /**
     * Deserializes from a buffer or reader, corresponding to a write in cpp.
     * @param buffer - Buffer or reader to read from.
     * @returns A new instance of PrivateKernelTailCircuitPublicInputs.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new PrivateKernelTailCircuitPublicInputs(reader.readObject(AggregationObject), reader.readObject(PrivateAccumulatedNonRevertibleData), reader.readObject(PrivateAccumulatedRevertibleData), reader.readObject(CombinedConstantData), reader.readBoolean(), reader.readBoolean(), reader.readBoolean());
    }
    static empty() {
        return new PrivateKernelTailCircuitPublicInputs(AggregationObject.makeFake(), PrivateAccumulatedNonRevertibleData.empty(), PrivateAccumulatedRevertibleData.empty(), CombinedConstantData.empty(), true, true, true);
    }
}
