// @aztec/circuits.js/dest/structs/kernel/private_kernel_tail_circuit_private_inputs.js
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { MAX_NEW_NOTE_HASHES_PER_TX, MAX_NEW_NULLIFIERS_PER_TX, MAX_NOTE_HASH_READ_REQUESTS_PER_TX, MAX_NULLIFIER_KEY_VALIDATION_REQUESTS_PER_TX, } from '../../constants.gen.js';
import { Fr, GrumpkinScalar } from '../index.js';
import { nullifierReadRequestHintsFromBuffer } from '../read_request_hints.js';
import { SideEffect, SideEffectLinkedToNoteHash } from '../side_effects.js';
import { PrivateKernelInnerData } from './private_kernel_inner_data.js';
/**
 * Input to the private kernel circuit - tail call.
 */
export class PrivateKernelTailCircuitPrivateInputs {
    constructor(
    /**
     * The previous kernel data
     */
    previousKernel, 
    /**
     * The sorted new note hashes.
     */
    sortedNewNoteHashes, 
    /**
     * The sorted new note hashes indexes. Maps original to sorted.
     */
    sortedNewNoteHashesIndexes, 
    /**
     * Contains hints for the transient read requests to localize corresponding commitments.
     */
    readCommitmentHints, 
    /**
     * The sorted new nullifiers. Maps original to sorted.
     */
    sortedNewNullifiers, 
    /**
     * The sorted new nullifiers indexes.
     */
    sortedNewNullifiersIndexes, 
    /**
     * Contains hints for the nullifier read requests to locate corresponding pending or settled nullifiers.
     */
    nullifierReadRequestHints, 
    /**
     * Contains hints for the transient nullifiers to localize corresponding commitments.
     */
    nullifierCommitmentHints, 
    /**
     * The master nullifier secret keys for the nullifier key validation requests.
     */
    masterNullifierSecretKeys) {
        this.previousKernel = previousKernel;
        this.sortedNewNoteHashes = sortedNewNoteHashes;
        this.sortedNewNoteHashesIndexes = sortedNewNoteHashesIndexes;
        this.readCommitmentHints = readCommitmentHints;
        this.sortedNewNullifiers = sortedNewNullifiers;
        this.sortedNewNullifiersIndexes = sortedNewNullifiersIndexes;
        this.nullifierReadRequestHints = nullifierReadRequestHints;
        this.nullifierCommitmentHints = nullifierCommitmentHints;
        this.masterNullifierSecretKeys = masterNullifierSecretKeys;
    }
    /**
     * Serialize this as a buffer.
     * @returns The buffer.
     */
    toBuffer() {
        return serializeToBuffer(this.previousKernel, this.sortedNewNoteHashes, this.sortedNewNoteHashesIndexes, this.readCommitmentHints, this.sortedNewNullifiers, this.sortedNewNullifiersIndexes, this.nullifierReadRequestHints, this.nullifierCommitmentHints, this.masterNullifierSecretKeys);
    }
    /**
     * Deserializes from a buffer or reader.
     * @param buffer - Buffer or reader to read from.
     * @returns The deserialized instance.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new PrivateKernelTailCircuitPrivateInputs(reader.readObject(PrivateKernelInnerData), reader.readArray(MAX_NEW_NOTE_HASHES_PER_TX, SideEffect), reader.readNumbers(MAX_NEW_NOTE_HASHES_PER_TX), reader.readArray(MAX_NOTE_HASH_READ_REQUESTS_PER_TX, Fr), reader.readArray(MAX_NEW_NULLIFIERS_PER_TX, SideEffectLinkedToNoteHash), reader.readNumbers(MAX_NEW_NULLIFIERS_PER_TX), reader.readObject({ fromBuffer: nullifierReadRequestHintsFromBuffer }), reader.readArray(MAX_NEW_NULLIFIERS_PER_TX, Fr), reader.readArray(MAX_NULLIFIER_KEY_VALIDATION_REQUESTS_PER_TX, GrumpkinScalar));
    }
}
