// @aztec/circuits.js/dest/structs/kernel/private_call_data.js
import { Fr } from '@aztec/foundation/fields';
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { FUNCTION_TREE_HEIGHT, MAX_NOTE_HASH_READ_REQUESTS_PER_CALL, MAX_PRIVATE_CALL_STACK_LENGTH_PER_CALL, MAX_PUBLIC_CALL_STACK_LENGTH_PER_CALL, } from '../../constants.gen.js';
import { CallRequest } from '../call_request.js';
import { MembershipWitness } from '../membership_witness.js';
import { NoteHashReadRequestMembershipWitness } from '../note_hash_read_request_membership_witness.js';
import { PrivateCallStackItem } from '../private_call_stack_item.js';
import { Proof } from '../proof.js';
import { VerificationKey } from '../verification_key.js';
/**
 * Private call data.
 */
export class PrivateCallData {
    constructor(
    /**
     * The call stack item currently being processed.
     */
    callStackItem, 
    /**
     * Other private call stack items to be processed.
     */
    privateCallStack, 
    /**
     * Other public call stack items to be processed.
     */
    publicCallStack, 
    /**
     * The proof of the execution of this private call.
     */
    proof, 
    /**
     * The verification key for the function being invoked.
     */
    vk, 
    /**
     * Artifact hash of the contract class for this private call.
     */
    contractClassArtifactHash, 
    /**
     * Public bytecode commitment for the contract class for this private call.
     */
    contractClassPublicBytecodeCommitment, 
    /**
     * Public keys hash of the contract instance.
     */
    publicKeysHash, 
    /**
     * Salted initialization hash of the contract instance.
     */
    saltedInitializationHash, 
    /**
     * The membership witness for the function leaf corresponding to the function being invoked.
     */
    functionLeafMembershipWitness, 
    /**
     * The membership witnesses for read requests created by the function being invoked.
     */
    noteHashReadRequestMembershipWitnesses, 
    /**
     * The address of the portal contract corresponding to the contract on which the function is being invoked.
     */
    portalContractAddress, 
    /**
     * The hash of the ACIR of the function being invoked.
     */
    acirHash) {
        this.callStackItem = callStackItem;
        this.privateCallStack = privateCallStack;
        this.publicCallStack = publicCallStack;
        this.proof = proof;
        this.vk = vk;
        this.contractClassArtifactHash = contractClassArtifactHash;
        this.contractClassPublicBytecodeCommitment = contractClassPublicBytecodeCommitment;
        this.publicKeysHash = publicKeysHash;
        this.saltedInitializationHash = saltedInitializationHash;
        this.functionLeafMembershipWitness = functionLeafMembershipWitness;
        this.noteHashReadRequestMembershipWitnesses = noteHashReadRequestMembershipWitnesses;
        this.portalContractAddress = portalContractAddress;
        this.acirHash = acirHash;
    }
    /**
     * Serialize into a field array. Low-level utility.
     * @param fields - Object with fields.
     * @returns The array.
     */
    static getFields(fields) {
        return [
            fields.callStackItem,
            fields.privateCallStack,
            fields.publicCallStack,
            fields.proof,
            fields.vk,
            fields.contractClassArtifactHash,
            fields.contractClassPublicBytecodeCommitment,
            fields.publicKeysHash,
            fields.saltedInitializationHash,
            fields.functionLeafMembershipWitness,
            fields.noteHashReadRequestMembershipWitnesses,
            fields.portalContractAddress,
            fields.acirHash,
        ];
    }
    static from(fields) {
        return new PrivateCallData(...PrivateCallData.getFields(fields));
    }
    /**
     * Serialize this as a buffer.
     * @returns The buffer.
     */
    toBuffer() {
        return serializeToBuffer(...PrivateCallData.getFields(this));
    }
    /**
     * Deserializes from a buffer or reader.
     * @param buffer - Buffer or reader to read from.
     * @returns The deserialized instance.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new PrivateCallData(reader.readObject(PrivateCallStackItem), reader.readArray(MAX_PRIVATE_CALL_STACK_LENGTH_PER_CALL, CallRequest), reader.readArray(MAX_PUBLIC_CALL_STACK_LENGTH_PER_CALL, CallRequest), reader.readObject(Proof), reader.readObject(VerificationKey), reader.readObject(Fr), reader.readObject(Fr), reader.readObject(Fr), reader.readObject(Fr), reader.readObject(MembershipWitness.deserializer(FUNCTION_TREE_HEIGHT)), reader.readArray(MAX_NOTE_HASH_READ_REQUESTS_PER_CALL, NoteHashReadRequestMembershipWitness), reader.readObject(Fr), reader.readObject(Fr));
    }
}
