// @aztec/circuits.js/dest/structs/kernel/public_call_data.js
import { serializeToBuffer } from '@aztec/foundation/serialize';
/**
 * Public calldata assembled from the kernel execution result and proof.
 */
export class PublicCallData {
    constructor(
    /**
     * Call stack item being processed by the current iteration of the kernel.
     */
    callStackItem, 
    /**
     * Children call stack items.
     */
    publicCallStack, 
    /**
     * Proof of the call stack item execution.
     */
    proof, 
    /**
     * Address of the corresponding portal contract.
     */
    portalContractAddress, 
    /**
     * Hash of the L2 contract bytecode.
     */
    bytecodeHash) {
        this.callStackItem = callStackItem;
        this.publicCallStack = publicCallStack;
        this.proof = proof;
        this.portalContractAddress = portalContractAddress;
        this.bytecodeHash = bytecodeHash;
    }
    toBuffer() {
        return serializeToBuffer(this.callStackItem, this.publicCallStack, this.proof, this.portalContractAddress, this.bytecodeHash);
    }
}
