// @aztec/types/dest/contracts/contract_instance.js
import { AztecAddress } from '@aztec/foundation/aztec-address';
import { EthAddress } from '@aztec/foundation/eth-address';
import { Fr } from '@aztec/foundation/fields';
import { BufferReader, numToUInt8, serializeToBuffer } from '@aztec/foundation/serialize';
const VERSION = 1;
export class SerializableContractInstance {
    constructor(instance) {
        this.version = VERSION;
        if (instance.version !== VERSION) {
            throw new Error(`Unexpected contract class version ${instance.version}`);
        }
        this.salt = instance.salt;
        this.contractClassId = instance.contractClassId;
        this.initializationHash = instance.initializationHash;
        this.portalContractAddress = instance.portalContractAddress;
        this.publicKeysHash = instance.publicKeysHash;
        this.deployer = instance.deployer;
    }
    toBuffer() {
        return serializeToBuffer(numToUInt8(this.version), this.salt, this.contractClassId, this.initializationHash, this.portalContractAddress, this.publicKeysHash, this.deployer);
    }
    /** Returns a copy of this object with its address included. */
    withAddress(address) {
        return { ...this, address };
    }
    static fromBuffer(bufferOrReader) {
        const reader = BufferReader.asReader(bufferOrReader);
        return new SerializableContractInstance({
            version: reader.readUInt8(),
            salt: reader.readObject(Fr),
            contractClassId: reader.readObject(Fr),
            initializationHash: reader.readObject(Fr),
            portalContractAddress: reader.readObject(EthAddress),
            publicKeysHash: reader.readObject(Fr),
            deployer: reader.readObject(AztecAddress),
        });
    }
    static random(opts = {}) {
        return new SerializableContractInstance({
            version: VERSION,
            salt: Fr.random(),
            contractClassId: Fr.random(),
            initializationHash: Fr.random(),
            portalContractAddress: EthAddress.random(),
            publicKeysHash: Fr.random(),
            deployer: AztecAddress.random(),
            ...opts,
        });
    }
}
