// @aztec/aztec.js/dest/fee/native_fee_payment_method.js
var _NativeFeePaymentMethod_gasTokenAddress;
import { __classPrivateFieldGet, __classPrivateFieldSet } from "tslib";
import { FunctionData } from '@aztec/circuits.js';
import { FunctionSelector } from '@aztec/foundation/abi';
import { getCanonicalGasTokenAddress } from '@aztec/protocol-contracts/gas-token';
/**
 * Pay fee directly in the native gas token.
 */
export class NativeFeePaymentMethod {
    constructor(canonicalGasTokenAddress) {
        _NativeFeePaymentMethod_gasTokenAddress.set(this, void 0);
        __classPrivateFieldSet(this, _NativeFeePaymentMethod_gasTokenAddress, canonicalGasTokenAddress, "f");
    }
    static async create(wallet) {
        const nodeInfo = await wallet.getNodeInfo();
        return new NativeFeePaymentMethod(getCanonicalGasTokenAddress(nodeInfo.l1ContractAddresses.gasPortalAddress));
    }
    /**
     * Gets the native gas asset used to pay the fee.
     * @returns The asset used to pay the fee.
     */
    getAsset() {
        return __classPrivateFieldGet(this, _NativeFeePaymentMethod_gasTokenAddress, "f");
    }
    /**
     * The contract responsible for fee payment. This will be the same as the asset.
     * @returns The contract address responsible for holding the fee payment.
     */
    getPaymentContract() {
        return __classPrivateFieldGet(this, _NativeFeePaymentMethod_gasTokenAddress, "f");
    }
    /**
     * Fee payments in the native gas token are always public.
     * @returns false
     */
    isPrivateFeePayment() {
        return false;
    }
    /**
     * Creates a function call to pay the fee in gas token..
     * @param feeLimit - The maximum fee to be paid in gas token.
     * @returns A function call
     */
    getFunctionCalls(feeLimit) {
        return Promise.resolve([
            {
                to: __classPrivateFieldGet(this, _NativeFeePaymentMethod_gasTokenAddress, "f"),
                functionData: new FunctionData(FunctionSelector.fromSignature('pay_fee(Field)'), false),
                args: [feeLimit],
            },
        ]);
    }
}
_NativeFeePaymentMethod_gasTokenAddress = new WeakMap();
