// @aztec/foundation/dest/abi/abi.js
import { inflate } from 'pako';
/**
 * Indicates whether a parameter is public or secret/private.
 */
export var ABIParameterVisibility;
(function (ABIParameterVisibility) {
    ABIParameterVisibility["PUBLIC"] = "public";
    ABIParameterVisibility["SECRET"] = "secret";
})(ABIParameterVisibility || (ABIParameterVisibility = {}));
/**
 * Aztec.nr function types.
 */
export var FunctionType;
(function (FunctionType) {
    FunctionType["SECRET"] = "secret";
    FunctionType["OPEN"] = "open";
    FunctionType["UNCONSTRAINED"] = "unconstrained";
})(FunctionType || (FunctionType = {}));
/**
 * Gets a function artifact given its name or selector.
 */
export function getFunctionArtifact(artifact, functionNameOrSelector) {
    const functionArtifact = artifact.functions.find(f => typeof functionNameOrSelector === 'string'
        ? f.name === functionNameOrSelector
        : functionNameOrSelector.equals(f.name, f.parameters));
    if (!functionArtifact) {
        throw new Error(`Unknown function ${functionNameOrSelector}`);
    }
    return functionArtifact;
}
/** @deprecated Use getFunctionArtifact instead */
export function getFunctionArtifactWithSelector(artifact, selector) {
    return getFunctionArtifact(artifact, selector);
}
/**
 * Gets a function artifact including debug metadata given its name or selector.
 */
export function getFunctionArtifactWithDebugMetadata(artifact, functionNameOrSelector) {
    const functionArtifact = getFunctionArtifact(artifact, functionNameOrSelector);
    const debugMetadata = getFunctionDebugMetadata(artifact, functionArtifact);
    return { ...functionArtifact, debug: debugMetadata };
}
/**
 * Gets the debug metadata of a given function from the contract artifact
 * @param artifact - The contract build artifact
 * @param functionName - The name of the function
 * @returns The debug metadata of the function
 */
export function getFunctionDebugMetadata(contractArtifact, functionArtifact) {
    if (functionArtifact.debugSymbols && contractArtifact.fileMap) {
        const debugSymbols = JSON.parse(inflate(Buffer.from(functionArtifact.debugSymbols, 'base64'), { to: 'string', raw: true }));
        return { debugSymbols, files: contractArtifact.fileMap };
    }
    return undefined;
}
/**
 * Returns an initializer from the contract, assuming there is at least one. If there are multiple initializers,
 * it returns the one named "constructor"; if there is none with that name, it returns the first private initializer
 * it finds.
 * @param contractArtifact - The contract artifact.
 * @returns An initializer function, or none if there are no functions flagged as initializers in the contract.
 */
export function getDefaultInitializer(contractArtifact) {
    const initializers = contractArtifact.functions.filter(f => f.isInitializer);
    return initializers.length > 1
        ? initializers.find(f => f.name === 'constructor') ??
            initializers.find(f => f.functionType === FunctionType.SECRET) ??
            initializers[0]
        : initializers[0];
}
