// @aztec/circuits.js/dest/structs/rollup/base_or_merge_rollup_public_inputs.js
import { Fr } from '@aztec/foundation/fields';
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { NUM_FIELDS_PER_SHA256 } from '../../constants.gen.js';
import { AggregationObject } from '../aggregation_object.js';
import { PartialStateReference } from '../partial_state_reference.js';
import { ConstantRollupData } from './base_rollup.js';
/**
 * Output of the base and merge rollup circuits.
 */
export class BaseOrMergeRollupPublicInputs {
    constructor(
    /**
     * Specifies from which type of rollup circuit these inputs are from.
     */
    rollupType, 
    /**
     * Rollup sub tree height.
     * Note 1: Base rollup circuit always have a sub tree height of 0.
     * Note 2: With each merge, the sub tree height increases by 1.
     */
    rollupSubtreeHeight, 
    /**
     * Native aggregation state at the end of the rollup circuit.
     */
    aggregationObject, 
    /**
     * Data which is forwarded through the rollup circuits unchanged.
     */
    constants, 
    /**
     * Partial state reference at the start of the rollup circuit.
     */
    start, 
    /**
     * Partial state reference at the end of the rollup circuit.
     */
    end, 
    /**
     * SHA256 hashes of transactions effects. Used to make public inputs constant-sized (to then be unpacked on-chain).
     * Note: Length 2 for high and low.
     */
    txsEffectsHash, 
    /**
     * SHA256 hashes of outhash. Used to make public inputs constant-sized (to then be unpacked on-chain).
     * Note: Length 2 for high and low.
     */
    outHash) {
        this.rollupType = rollupType;
        this.rollupSubtreeHeight = rollupSubtreeHeight;
        this.aggregationObject = aggregationObject;
        this.constants = constants;
        this.start = start;
        this.end = end;
        this.txsEffectsHash = txsEffectsHash;
        this.outHash = outHash;
    }
    /**
     * Deserializes from a buffer or reader.
     * Note: Corresponds to a write in cpp.
     * @param buffer - Buffer or reader to read from.
     * @returns The deserialized public inputs.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new BaseOrMergeRollupPublicInputs(reader.readNumber(), Fr.fromBuffer(reader), reader.readObject(AggregationObject), reader.readObject(ConstantRollupData), reader.readObject(PartialStateReference), reader.readObject(PartialStateReference), reader.readArray(NUM_FIELDS_PER_SHA256, Fr), reader.readArray(NUM_FIELDS_PER_SHA256, Fr));
    }
    /**
     * Serialize this as a buffer.
     * @returns The buffer.
     */
    toBuffer() {
        return serializeToBuffer(this.rollupType, this.rollupSubtreeHeight, this.aggregationObject, this.constants, this.start, this.end, this.txsEffectsHash, this.outHash);
    }
}
