// @aztec/circuits.js/dest/structs/rollup/base_rollup.js
import { Fr } from '@aztec/foundation/fields';
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { GlobalVariables } from '../global_variables.js';
import { AppendOnlyTreeSnapshot } from './append_only_tree_snapshot.js';
import { NullifierLeaf, NullifierLeafPreimage } from './nullifier_leaf/index.js';
import { PublicDataTreeLeaf, PublicDataTreeLeafPreimage } from './public_data_leaf/index.js';
export { NullifierLeaf, NullifierLeafPreimage, PublicDataTreeLeaf, PublicDataTreeLeafPreimage };
/**
 * Data which is forwarded through the base rollup circuits unchanged.
 */
export class ConstantRollupData {
    constructor(
    /** Archive tree snapshot at the very beginning of the entire rollup. */
    lastArchive, 
    /**
     * Root of the private kernel verification key tree.
     */
    privateKernelVkTreeRoot, 
    /**
     * Root of the public kernel circuit verification key tree.
     */
    publicKernelVkTreeRoot, 
    /**
     * Hash of the base rollup circuit verification key.
     */
    baseRollupVkHash, 
    /**
     * Hash of the merge rollup circuit verification key.
     */
    mergeRollupVkHash, 
    /**
     * Global variables for the block
     */
    globalVariables) {
        this.lastArchive = lastArchive;
        this.privateKernelVkTreeRoot = privateKernelVkTreeRoot;
        this.publicKernelVkTreeRoot = publicKernelVkTreeRoot;
        this.baseRollupVkHash = baseRollupVkHash;
        this.mergeRollupVkHash = mergeRollupVkHash;
        this.globalVariables = globalVariables;
    }
    static from(fields) {
        return new ConstantRollupData(...ConstantRollupData.getFields(fields));
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new ConstantRollupData(reader.readObject(AppendOnlyTreeSnapshot), Fr.fromBuffer(reader), Fr.fromBuffer(reader), Fr.fromBuffer(reader), Fr.fromBuffer(reader), reader.readObject(GlobalVariables));
    }
    static getFields(fields) {
        return [
            fields.lastArchive,
            fields.privateKernelVkTreeRoot,
            fields.publicKernelVkTreeRoot,
            fields.baseRollupVkHash,
            fields.mergeRollupVkHash,
            fields.globalVariables,
        ];
    }
    toBuffer() {
        return serializeToBuffer(...ConstantRollupData.getFields(this));
    }
}
/**
 * Inputs to the base rollup circuit.
 */
export class BaseRollupInputs {
    constructor(
    /** Data of the 2 kernels that preceded this base rollup circuit. */
    kernelData, 
    /** Partial state reference at the start of the rollup. */
    start, 
    /** Hints used while proving state diff validity. */
    stateDiffHints, 
    /**
     * The public data writes to be inserted in the tree, sorted high slot to low slot.
     */
    sortedPublicDataWrites, 
    /**
     * The indexes of the sorted public data writes to the original ones.
     */
    sortedPublicDataWritesIndexes, 
    /**
     * The public data writes which need to be updated to perform the batch insertion of the new public data writes.
     * See `StandardIndexedTree.batchInsert` function for more details.
     */
    lowPublicDataWritesPreimages, 
    /**
     * Membership witnesses for the nullifiers which need to be updated to perform the batch insertion of the new
     * nullifiers.
     */
    lowPublicDataWritesMembershipWitnesses, 
    /**
     * Preimages of leaves which are to be read by the public data reads.
     */
    publicDataReadsPreimages, 
    /**
     * Sibling paths of leaves which are to be read by the public data reads.
     * Each item in the array is the sibling path that corresponds to a read request.
     */
    publicDataReadsMembershipWitnesses, 
    /**
     * Membership witnesses of blocks referred by each of the 2 kernels.
     */
    archiveRootMembershipWitness, 
    /**
     * Data which is not modified by the base rollup circuit.
     */
    constants) {
        this.kernelData = kernelData;
        this.start = start;
        this.stateDiffHints = stateDiffHints;
        this.sortedPublicDataWrites = sortedPublicDataWrites;
        this.sortedPublicDataWritesIndexes = sortedPublicDataWritesIndexes;
        this.lowPublicDataWritesPreimages = lowPublicDataWritesPreimages;
        this.lowPublicDataWritesMembershipWitnesses = lowPublicDataWritesMembershipWitnesses;
        this.publicDataReadsPreimages = publicDataReadsPreimages;
        this.publicDataReadsMembershipWitnesses = publicDataReadsMembershipWitnesses;
        this.archiveRootMembershipWitness = archiveRootMembershipWitness;
        this.constants = constants;
    }
    static from(fields) {
        return new BaseRollupInputs(...BaseRollupInputs.getFields(fields));
    }
    static getFields(fields) {
        return [
            fields.kernelData,
            fields.start,
            fields.stateDiffHints,
            fields.sortedPublicDataWrites,
            fields.sortedPublicDataWritesIndexes,
            fields.lowPublicDataWritesPreimages,
            fields.lowPublicDataWritesMembershipWitnesses,
            fields.publicDataReadsPreimages,
            fields.publicDataReadsMembershipWitnesses,
            fields.archiveRootMembershipWitness,
            fields.constants,
        ];
    }
    toBuffer() {
        return serializeToBuffer(...BaseRollupInputs.getFields(this));
    }
}
