// @aztec/noir-contracts.js/src/AppSubscription.ts
/* Autogenerated file, do not edit! */

/* eslint-disable */
import {
  AztecAddress,
  AztecAddressLike,
  CompleteAddress,
  Contract,
  ContractArtifact,
  ContractBase,
  ContractFunctionInteraction,
  ContractInstanceWithAddress,
  ContractMethod,
  DeployMethod,
  EthAddress,
  EthAddressLike,
  FieldLike,
  Fr,
  FunctionSelectorLike,
  NoirCompiledContract,
  Point,
  PublicKey,
  Wallet,
  WrappedFieldLike,
  loadContractArtifact,
} from '@aztec/aztec.js';

import AppSubscriptionContractArtifactJson from '../artifacts/app_subscription_contract-AppSubscription.json' assert { type: 'json' };

export const AppSubscriptionContractArtifact = loadContractArtifact(
  AppSubscriptionContractArtifactJson as NoirCompiledContract,
);

/**
 * Type-safe interface for contract AppSubscription;
 */
export class AppSubscriptionContract extends ContractBase {
  private constructor(instance: ContractInstanceWithAddress, wallet: Wallet) {
    super(instance, AppSubscriptionContractArtifact, wallet);
  }

  /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */
  public static async at(address: AztecAddress, wallet: Wallet) {
    return Contract.at(address, AppSubscriptionContract.artifact, wallet) as Promise<AppSubscriptionContract>;
  }

  /**
   * Creates a tx to deploy a new instance of this contract.
   */
  public static deploy(
    wallet: Wallet,
    target_address: AztecAddressLike,
    subscription_recipient_address: AztecAddressLike,
    subscription_token_address: AztecAddressLike,
    subscription_price: FieldLike,
    gas_token_address: AztecAddressLike,
  ) {
    return new DeployMethod<AppSubscriptionContract>(
      Point.ZERO,
      wallet,
      AppSubscriptionContractArtifact,
      AppSubscriptionContract.at,
      Array.from(arguments).slice(1),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified public key to derive the address.
   */
  public static deployWithPublicKey(
    publicKey: PublicKey,
    wallet: Wallet,
    target_address: AztecAddressLike,
    subscription_recipient_address: AztecAddressLike,
    subscription_token_address: AztecAddressLike,
    subscription_price: FieldLike,
    gas_token_address: AztecAddressLike,
  ) {
    return new DeployMethod<AppSubscriptionContract>(
      publicKey,
      wallet,
      AppSubscriptionContractArtifact,
      AppSubscriptionContract.at,
      Array.from(arguments).slice(2),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */
  public static deployWithOpts<M extends keyof AppSubscriptionContract['methods']>(
    opts: { publicKey?: PublicKey; method?: M; wallet: Wallet },
    ...args: Parameters<AppSubscriptionContract['methods'][M]>
  ) {
    return new DeployMethod<AppSubscriptionContract>(
      opts.publicKey ?? Point.ZERO,
      opts.wallet,
      AppSubscriptionContractArtifact,
      AppSubscriptionContract.at,
      Array.from(arguments).slice(1),
      opts.method ?? 'constructor',
    );
  }

  /**
   * Returns this contract's artifact.
   */
  public static get artifact(): ContractArtifact {
    return AppSubscriptionContractArtifact;
  }

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public methods!: {
    /** is_initialized(subscriber_address: struct) */
    is_initialized: ((subscriber_address: AztecAddressLike) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** entrypoint(payload: struct, user_address: struct) */
    entrypoint: ((
      payload: {
        function_calls: {
          args_hash: FieldLike;
          function_selector: FunctionSelectorLike;
          target_address: AztecAddressLike;
          is_public: boolean;
        }[];
        nonce: FieldLike;
      },
      user_address: AztecAddressLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** constructor(target_address: struct, subscription_recipient_address: struct, subscription_token_address: struct, subscription_price: field, gas_token_address: struct) */
    constructor: ((
      target_address: AztecAddressLike,
      subscription_recipient_address: AztecAddressLike,
      subscription_token_address: AztecAddressLike,
      subscription_price: FieldLike,
      gas_token_address: AztecAddressLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** compute_note_hash_and_nullifier(contract_address: struct, nonce: field, storage_slot: field, note_type_id: field, serialized_note: array) */
    compute_note_hash_and_nullifier: ((
      contract_address: AztecAddressLike,
      nonce: FieldLike,
      storage_slot: FieldLike,
      note_type_id: FieldLike,
      serialized_note: FieldLike[],
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** subscribe(subscriber_address: struct, nonce: field, expiry_block_number: field, tx_count: field) */
    subscribe: ((
      subscriber_address: AztecAddressLike,
      nonce: FieldLike,
      expiry_block_number: FieldLike,
      tx_count: FieldLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;
  };
}
