// @aztec/noir-contracts.js/src/Child.ts
/* Autogenerated file, do not edit! */

/* eslint-disable */
import {
  AztecAddress,
  AztecAddressLike,
  CompleteAddress,
  Contract,
  ContractArtifact,
  ContractBase,
  ContractFunctionInteraction,
  ContractInstanceWithAddress,
  ContractMethod,
  DeployMethod,
  EthAddress,
  EthAddressLike,
  FieldLike,
  Fr,
  FunctionSelectorLike,
  NoirCompiledContract,
  Point,
  PublicKey,
  Wallet,
  WrappedFieldLike,
  loadContractArtifact,
} from '@aztec/aztec.js';

import ChildContractArtifactJson from '../artifacts/child_contract-Child.json' assert { type: 'json' };

export const ChildContractArtifact = loadContractArtifact(ChildContractArtifactJson as NoirCompiledContract);

/**
 * Type-safe interface for contract Child;
 */
export class ChildContract extends ContractBase {
  private constructor(instance: ContractInstanceWithAddress, wallet: Wallet) {
    super(instance, ChildContractArtifact, wallet);
  }

  /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */
  public static async at(address: AztecAddress, wallet: Wallet) {
    return Contract.at(address, ChildContract.artifact, wallet) as Promise<ChildContract>;
  }

  /**
   * Creates a tx to deploy a new instance of this contract.
   */
  public static deploy(wallet: Wallet) {
    return new DeployMethod<ChildContract>(
      Point.ZERO,
      wallet,
      ChildContractArtifact,
      ChildContract.at,
      Array.from(arguments).slice(1),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified public key to derive the address.
   */
  public static deployWithPublicKey(publicKey: PublicKey, wallet: Wallet) {
    return new DeployMethod<ChildContract>(
      publicKey,
      wallet,
      ChildContractArtifact,
      ChildContract.at,
      Array.from(arguments).slice(2),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */
  public static deployWithOpts<M extends keyof ChildContract['methods']>(
    opts: { publicKey?: PublicKey; method?: M; wallet: Wallet },
    ...args: Parameters<ChildContract['methods'][M]>
  ) {
    return new DeployMethod<ChildContract>(
      opts.publicKey ?? Point.ZERO,
      opts.wallet,
      ChildContractArtifact,
      ChildContract.at,
      Array.from(arguments).slice(1),
      opts.method ?? 'constructor',
    );
  }

  /**
   * Returns this contract's artifact.
   */
  public static get artifact(): ContractArtifact {
    return ChildContractArtifact;
  }

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public methods!: {
    /** value(input: field) */
    value: ((input: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** compute_note_hash_and_nullifier(contract_address: struct, nonce: field, storage_slot: field, note_type_id: field, serialized_note: array) */
    compute_note_hash_and_nullifier: ((
      contract_address: AztecAddressLike,
      nonce: FieldLike,
      storage_slot: FieldLike,
      note_type_id: FieldLike,
      serialized_note: FieldLike[],
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** setValueTwiceWithNestedLast() */
    setValueTwiceWithNestedLast: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** setValueTwiceWithNestedFirst() */
    setValueTwiceWithNestedFirst: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** pubSetValue(new_value: field) */
    pubSetValue: ((new_value: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** pubGetValue(base_value: field) */
    pubGetValue: ((base_value: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** pubIncValue(new_value: field) */
    pubIncValue: ((new_value: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** privateSetValue(new_value: field, owner: struct) */
    privateSetValue: ((new_value: FieldLike, owner: AztecAddressLike) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** privateGetValue(amount: field, owner: struct) */
    privateGetValue: ((amount: FieldLike, owner: AztecAddressLike) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;
  };
}
