// @aztec/noir-contracts.js/src/ContractInstanceDeployer.ts
/* Autogenerated file, do not edit! */

/* eslint-disable */
import {
  AztecAddress,
  AztecAddressLike,
  CompleteAddress,
  Contract,
  ContractArtifact,
  ContractBase,
  ContractFunctionInteraction,
  ContractInstanceWithAddress,
  ContractMethod,
  DeployMethod,
  EthAddress,
  EthAddressLike,
  FieldLike,
  Fr,
  FunctionSelectorLike,
  NoirCompiledContract,
  Point,
  PublicKey,
  Wallet,
  WrappedFieldLike,
  loadContractArtifact,
} from '@aztec/aztec.js';

import ContractInstanceDeployerContractArtifactJson from '../artifacts/contract_instance_deployer_contract-ContractInstanceDeployer.json' assert { type: 'json' };

export const ContractInstanceDeployerContractArtifact = loadContractArtifact(
  ContractInstanceDeployerContractArtifactJson as NoirCompiledContract,
);

/**
 * Type-safe interface for contract ContractInstanceDeployer;
 */
export class ContractInstanceDeployerContract extends ContractBase {
  private constructor(instance: ContractInstanceWithAddress, wallet: Wallet) {
    super(instance, ContractInstanceDeployerContractArtifact, wallet);
  }

  /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */
  public static async at(address: AztecAddress, wallet: Wallet) {
    return Contract.at(
      address,
      ContractInstanceDeployerContract.artifact,
      wallet,
    ) as Promise<ContractInstanceDeployerContract>;
  }

  /**
   * Creates a tx to deploy a new instance of this contract.
   */
  public static deploy(wallet: Wallet) {
    return new DeployMethod<ContractInstanceDeployerContract>(
      Point.ZERO,
      wallet,
      ContractInstanceDeployerContractArtifact,
      ContractInstanceDeployerContract.at,
      Array.from(arguments).slice(1),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified public key to derive the address.
   */
  public static deployWithPublicKey(publicKey: PublicKey, wallet: Wallet) {
    return new DeployMethod<ContractInstanceDeployerContract>(
      publicKey,
      wallet,
      ContractInstanceDeployerContractArtifact,
      ContractInstanceDeployerContract.at,
      Array.from(arguments).slice(2),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */
  public static deployWithOpts<M extends keyof ContractInstanceDeployerContract['methods']>(
    opts: { publicKey?: PublicKey; method?: M; wallet: Wallet },
    ...args: Parameters<ContractInstanceDeployerContract['methods'][M]>
  ) {
    return new DeployMethod<ContractInstanceDeployerContract>(
      opts.publicKey ?? Point.ZERO,
      opts.wallet,
      ContractInstanceDeployerContractArtifact,
      ContractInstanceDeployerContract.at,
      Array.from(arguments).slice(1),
      opts.method ?? 'constructor',
    );
  }

  /**
   * Returns this contract's artifact.
   */
  public static get artifact(): ContractArtifact {
    return ContractInstanceDeployerContractArtifact;
  }

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public methods!: {
    /** compute_note_hash_and_nullifier(contract_address: struct, nonce: field, storage_slot: field, note_type_id: field, serialized_note: array) */
    compute_note_hash_and_nullifier: ((
      contract_address: AztecAddressLike,
      nonce: FieldLike,
      storage_slot: FieldLike,
      note_type_id: FieldLike,
      serialized_note: FieldLike[],
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** deploy(salt: field, contract_class_id: struct, initialization_hash: field, portal_contract_address: struct, public_keys_hash: struct, universal_deploy: boolean) */
    deploy: ((
      salt: FieldLike,
      contract_class_id: WrappedFieldLike,
      initialization_hash: FieldLike,
      portal_contract_address: EthAddressLike,
      public_keys_hash: WrappedFieldLike,
      universal_deploy: boolean,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;
  };
}
