// @aztec/noir-contracts.js/src/DelegatedOn.ts
/* Autogenerated file, do not edit! */

/* eslint-disable */
import {
  AztecAddress,
  AztecAddressLike,
  CompleteAddress,
  Contract,
  ContractArtifact,
  ContractBase,
  ContractFunctionInteraction,
  ContractInstanceWithAddress,
  ContractMethod,
  DeployMethod,
  EthAddress,
  EthAddressLike,
  FieldLike,
  Fr,
  FunctionSelectorLike,
  NoirCompiledContract,
  Point,
  PublicKey,
  Wallet,
  WrappedFieldLike,
  loadContractArtifact,
} from '@aztec/aztec.js';

import DelegatedOnContractArtifactJson from '../artifacts/delegated_on_contract-DelegatedOn.json' assert { type: 'json' };

export const DelegatedOnContractArtifact = loadContractArtifact(
  DelegatedOnContractArtifactJson as NoirCompiledContract,
);

/**
 * Type-safe interface for contract DelegatedOn;
 */
export class DelegatedOnContract extends ContractBase {
  private constructor(instance: ContractInstanceWithAddress, wallet: Wallet) {
    super(instance, DelegatedOnContractArtifact, wallet);
  }

  /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */
  public static async at(address: AztecAddress, wallet: Wallet) {
    return Contract.at(address, DelegatedOnContract.artifact, wallet) as Promise<DelegatedOnContract>;
  }

  /**
   * Creates a tx to deploy a new instance of this contract.
   */
  public static deploy(wallet: Wallet) {
    return new DeployMethod<DelegatedOnContract>(
      Point.ZERO,
      wallet,
      DelegatedOnContractArtifact,
      DelegatedOnContract.at,
      Array.from(arguments).slice(1),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified public key to derive the address.
   */
  public static deployWithPublicKey(publicKey: PublicKey, wallet: Wallet) {
    return new DeployMethod<DelegatedOnContract>(
      publicKey,
      wallet,
      DelegatedOnContractArtifact,
      DelegatedOnContract.at,
      Array.from(arguments).slice(2),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */
  public static deployWithOpts<M extends keyof DelegatedOnContract['methods']>(
    opts: { publicKey?: PublicKey; method?: M; wallet: Wallet },
    ...args: Parameters<DelegatedOnContract['methods'][M]>
  ) {
    return new DeployMethod<DelegatedOnContract>(
      opts.publicKey ?? Point.ZERO,
      opts.wallet,
      DelegatedOnContractArtifact,
      DelegatedOnContract.at,
      Array.from(arguments).slice(1),
      opts.method ?? 'constructor',
    );
  }

  /**
   * Returns this contract's artifact.
   */
  public static get artifact(): ContractArtifact {
    return DelegatedOnContractArtifact;
  }

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public methods!: {
    /** view_private_value(amount: field, owner: struct) */
    view_private_value: ((amount: FieldLike, owner: AztecAddressLike) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** private_set_value(new_value: field, owner: struct) */
    private_set_value: ((new_value: FieldLike, owner: AztecAddressLike) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** public_set_value(new_value: field) */
    public_set_value: ((new_value: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** compute_note_hash_and_nullifier(contract_address: struct, nonce: field, storage_slot: field, note_type_id: field, serialized_note: array) */
    compute_note_hash_and_nullifier: ((
      contract_address: AztecAddressLike,
      nonce: FieldLike,
      storage_slot: FieldLike,
      note_type_id: FieldLike,
      serialized_note: FieldLike[],
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** view_public_value() */
    view_public_value: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;
  };
}
