// @aztec/noir-contracts.js/src/Delegator.ts
/* Autogenerated file, do not edit! */

/* eslint-disable */
import {
  AztecAddress,
  AztecAddressLike,
  CompleteAddress,
  Contract,
  ContractArtifact,
  ContractBase,
  ContractFunctionInteraction,
  ContractInstanceWithAddress,
  ContractMethod,
  DeployMethod,
  EthAddress,
  EthAddressLike,
  FieldLike,
  Fr,
  FunctionSelectorLike,
  NoirCompiledContract,
  Point,
  PublicKey,
  Wallet,
  WrappedFieldLike,
  loadContractArtifact,
} from '@aztec/aztec.js';

import DelegatorContractArtifactJson from '../artifacts/delegator_contract-Delegator.json' assert { type: 'json' };

export const DelegatorContractArtifact = loadContractArtifact(DelegatorContractArtifactJson as NoirCompiledContract);

/**
 * Type-safe interface for contract Delegator;
 */
export class DelegatorContract extends ContractBase {
  private constructor(instance: ContractInstanceWithAddress, wallet: Wallet) {
    super(instance, DelegatorContractArtifact, wallet);
  }

  /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */
  public static async at(address: AztecAddress, wallet: Wallet) {
    return Contract.at(address, DelegatorContract.artifact, wallet) as Promise<DelegatorContract>;
  }

  /**
   * Creates a tx to deploy a new instance of this contract.
   */
  public static deploy(wallet: Wallet) {
    return new DeployMethod<DelegatorContract>(
      Point.ZERO,
      wallet,
      DelegatorContractArtifact,
      DelegatorContract.at,
      Array.from(arguments).slice(1),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified public key to derive the address.
   */
  public static deployWithPublicKey(publicKey: PublicKey, wallet: Wallet) {
    return new DeployMethod<DelegatorContract>(
      publicKey,
      wallet,
      DelegatorContractArtifact,
      DelegatorContract.at,
      Array.from(arguments).slice(2),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */
  public static deployWithOpts<M extends keyof DelegatorContract['methods']>(
    opts: { publicKey?: PublicKey; method?: M; wallet: Wallet },
    ...args: Parameters<DelegatorContract['methods'][M]>
  ) {
    return new DeployMethod<DelegatorContract>(
      opts.publicKey ?? Point.ZERO,
      opts.wallet,
      DelegatorContractArtifact,
      DelegatorContract.at,
      Array.from(arguments).slice(1),
      opts.method ?? 'constructor',
    );
  }

  /**
   * Returns this contract's artifact.
   */
  public static get artifact(): ContractArtifact {
    return DelegatorContractArtifact;
  }

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public methods!: {
    /** enqueued_delegate_set_value(targetContract: struct, targetSelector: struct, args: array) */
    enqueued_delegate_set_value: ((
      targetContract: AztecAddressLike,
      targetSelector: FunctionSelectorLike,
      args: FieldLike[],
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** public_delegate_set_value(targetContract: struct, targetSelector: struct, args: array) */
    public_delegate_set_value: ((
      targetContract: AztecAddressLike,
      targetSelector: FunctionSelectorLike,
      args: FieldLike[],
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** view_private_value(amount: field, owner: struct) */
    view_private_value: ((amount: FieldLike, owner: AztecAddressLike) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** view_public_value() */
    view_public_value: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** private_delegate_set_value(targetContract: struct, targetSelector: struct, args: array) */
    private_delegate_set_value: ((
      targetContract: AztecAddressLike,
      targetSelector: FunctionSelectorLike,
      args: FieldLike[],
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** compute_note_hash_and_nullifier(contract_address: struct, nonce: field, storage_slot: field, note_type_id: field, serialized_note: array) */
    compute_note_hash_and_nullifier: ((
      contract_address: AztecAddressLike,
      nonce: FieldLike,
      storage_slot: FieldLike,
      note_type_id: FieldLike,
      serialized_note: FieldLike[],
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;
  };
}
