// @aztec/noir-contracts.js/src/FPC.ts
/* Autogenerated file, do not edit! */

/* eslint-disable */
import {
  AztecAddress,
  AztecAddressLike,
  CompleteAddress,
  Contract,
  ContractArtifact,
  ContractBase,
  ContractFunctionInteraction,
  ContractInstanceWithAddress,
  ContractMethod,
  DeployMethod,
  EthAddress,
  EthAddressLike,
  FieldLike,
  Fr,
  FunctionSelectorLike,
  NoirCompiledContract,
  Point,
  PublicKey,
  Wallet,
  WrappedFieldLike,
  loadContractArtifact,
} from '@aztec/aztec.js';

import FPCContractArtifactJson from '../artifacts/fpc_contract-FPC.json' assert { type: 'json' };

export const FPCContractArtifact = loadContractArtifact(FPCContractArtifactJson as NoirCompiledContract);

/**
 * Type-safe interface for contract FPC;
 */
export class FPCContract extends ContractBase {
  private constructor(instance: ContractInstanceWithAddress, wallet: Wallet) {
    super(instance, FPCContractArtifact, wallet);
  }

  /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */
  public static async at(address: AztecAddress, wallet: Wallet) {
    return Contract.at(address, FPCContract.artifact, wallet) as Promise<FPCContract>;
  }

  /**
   * Creates a tx to deploy a new instance of this contract.
   */
  public static deploy(wallet: Wallet, other_asset: AztecAddressLike, gas_token_address: AztecAddressLike) {
    return new DeployMethod<FPCContract>(
      Point.ZERO,
      wallet,
      FPCContractArtifact,
      FPCContract.at,
      Array.from(arguments).slice(1),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified public key to derive the address.
   */
  public static deployWithPublicKey(
    publicKey: PublicKey,
    wallet: Wallet,
    other_asset: AztecAddressLike,
    gas_token_address: AztecAddressLike,
  ) {
    return new DeployMethod<FPCContract>(
      publicKey,
      wallet,
      FPCContractArtifact,
      FPCContract.at,
      Array.from(arguments).slice(2),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */
  public static deployWithOpts<M extends keyof FPCContract['methods']>(
    opts: { publicKey?: PublicKey; method?: M; wallet: Wallet },
    ...args: Parameters<FPCContract['methods'][M]>
  ) {
    return new DeployMethod<FPCContract>(
      opts.publicKey ?? Point.ZERO,
      opts.wallet,
      FPCContractArtifact,
      FPCContract.at,
      Array.from(arguments).slice(1),
      opts.method ?? 'constructor',
    );
  }

  /**
   * Returns this contract's artifact.
   */
  public static get artifact(): ContractArtifact {
    return FPCContractArtifact;
  }

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public methods!: {
    /** compute_note_hash_and_nullifier(contract_address: struct, nonce: field, storage_slot: field, note_type_id: field, serialized_note: array) */
    compute_note_hash_and_nullifier: ((
      contract_address: AztecAddressLike,
      nonce: FieldLike,
      storage_slot: FieldLike,
      note_type_id: FieldLike,
      serialized_note: FieldLike[],
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** constructor(other_asset: struct, gas_token_address: struct) */
    constructor: ((other_asset: AztecAddressLike, gas_token_address: AztecAddressLike) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** fee_entrypoint_private(amount: field, asset: struct, secret_hash: field, nonce: field) */
    fee_entrypoint_private: ((
      amount: FieldLike,
      asset: AztecAddressLike,
      secret_hash: FieldLike,
      nonce: FieldLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** fee_entrypoint_public(amount: field, asset: struct, nonce: field) */
    fee_entrypoint_public: ((
      amount: FieldLike,
      asset: AztecAddressLike,
      nonce: FieldLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;
  };
}
