// @aztec/noir-contracts.js/src/PendingNoteHashes.ts
/* Autogenerated file, do not edit! */

/* eslint-disable */
import {
  AztecAddress,
  AztecAddressLike,
  CompleteAddress,
  Contract,
  ContractArtifact,
  ContractBase,
  ContractFunctionInteraction,
  ContractInstanceWithAddress,
  ContractMethod,
  DeployMethod,
  EthAddress,
  EthAddressLike,
  FieldLike,
  Fr,
  FunctionSelectorLike,
  NoirCompiledContract,
  Point,
  PublicKey,
  Wallet,
  WrappedFieldLike,
  loadContractArtifact,
} from '@aztec/aztec.js';

import PendingNoteHashesContractArtifactJson from '../artifacts/pending_note_hashes_contract-PendingNoteHashes.json' assert { type: 'json' };

export const PendingNoteHashesContractArtifact = loadContractArtifact(
  PendingNoteHashesContractArtifactJson as NoirCompiledContract,
);

/**
 * Type-safe interface for contract PendingNoteHashes;
 */
export class PendingNoteHashesContract extends ContractBase {
  private constructor(instance: ContractInstanceWithAddress, wallet: Wallet) {
    super(instance, PendingNoteHashesContractArtifact, wallet);
  }

  /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */
  public static async at(address: AztecAddress, wallet: Wallet) {
    return Contract.at(address, PendingNoteHashesContract.artifact, wallet) as Promise<PendingNoteHashesContract>;
  }

  /**
   * Creates a tx to deploy a new instance of this contract.
   */
  public static deploy(wallet: Wallet) {
    return new DeployMethod<PendingNoteHashesContract>(
      Point.ZERO,
      wallet,
      PendingNoteHashesContractArtifact,
      PendingNoteHashesContract.at,
      Array.from(arguments).slice(1),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified public key to derive the address.
   */
  public static deployWithPublicKey(publicKey: PublicKey, wallet: Wallet) {
    return new DeployMethod<PendingNoteHashesContract>(
      publicKey,
      wallet,
      PendingNoteHashesContractArtifact,
      PendingNoteHashesContract.at,
      Array.from(arguments).slice(2),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */
  public static deployWithOpts<M extends keyof PendingNoteHashesContract['methods']>(
    opts: { publicKey?: PublicKey; method?: M; wallet: Wallet },
    ...args: Parameters<PendingNoteHashesContract['methods'][M]>
  ) {
    return new DeployMethod<PendingNoteHashesContract>(
      opts.publicKey ?? Point.ZERO,
      opts.wallet,
      PendingNoteHashesContractArtifact,
      PendingNoteHashesContract.at,
      Array.from(arguments).slice(1),
      opts.method ?? 'constructor',
    );
  }

  /**
   * Returns this contract's artifact.
   */
  public static get artifact(): ContractArtifact {
    return PendingNoteHashesContractArtifact;
  }

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public methods!: {
    /** insert_note(amount: field, owner: struct) */
    insert_note: ((amount: FieldLike, owner: AztecAddressLike) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** dummy(amount: field, owner: struct) */
    dummy: ((amount: FieldLike, owner: AztecAddressLike) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** test_bad_get_then_insert_flat(amount: field, owner: struct) */
    test_bad_get_then_insert_flat: ((amount: FieldLike, owner: AztecAddressLike) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** test_insert2_then_get2_then_nullify2_all_in_nested_calls(amount: field, owner: struct, insert_fn_selector: struct, get_then_nullify_fn_selector: struct) */
    test_insert2_then_get2_then_nullify2_all_in_nested_calls: ((
      amount: FieldLike,
      owner: AztecAddressLike,
      insert_fn_selector: FunctionSelectorLike,
      get_then_nullify_fn_selector: FunctionSelectorLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** test_insert_then_get_then_nullify_flat(amount: field, owner: struct) */
    test_insert_then_get_then_nullify_flat: ((
      amount: FieldLike,
      owner: AztecAddressLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** get_then_nullify_note(expected_value: field, owner: struct) */
    get_then_nullify_note: ((expected_value: FieldLike, owner: AztecAddressLike) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** test_insert_then_get_then_nullify_all_in_nested_calls(amount: field, owner: struct, insert_fn_selector: struct, get_then_nullify_fn_selector: struct, get_note_zero_fn_selector: struct) */
    test_insert_then_get_then_nullify_all_in_nested_calls: ((
      amount: FieldLike,
      owner: AztecAddressLike,
      insert_fn_selector: FunctionSelectorLike,
      get_then_nullify_fn_selector: FunctionSelectorLike,
      get_note_zero_fn_selector: FunctionSelectorLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** test_insert2_then_get2_then_nullify1_all_in_nested_calls(amount: field, owner: struct, insert_fn_selector: struct, get_then_nullify_fn_selector: struct) */
    test_insert2_then_get2_then_nullify1_all_in_nested_calls: ((
      amount: FieldLike,
      owner: AztecAddressLike,
      insert_fn_selector: FunctionSelectorLike,
      get_then_nullify_fn_selector: FunctionSelectorLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** test_insert1_then_get2_then_nullify2_all_in_nested_calls(amount: field, owner: struct, insert_fn_selector: struct, get_then_nullify_fn_selector: struct, get_note_zero_fn_selector: struct) */
    test_insert1_then_get2_then_nullify2_all_in_nested_calls: ((
      amount: FieldLike,
      owner: AztecAddressLike,
      insert_fn_selector: FunctionSelectorLike,
      get_then_nullify_fn_selector: FunctionSelectorLike,
      get_note_zero_fn_selector: FunctionSelectorLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** get_note_zero_balance(owner: struct) */
    get_note_zero_balance: ((owner: AztecAddressLike) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** compute_note_hash_and_nullifier(contract_address: struct, nonce: field, storage_slot: field, note_type_id: field, serialized_note: array) */
    compute_note_hash_and_nullifier: ((
      contract_address: AztecAddressLike,
      nonce: FieldLike,
      storage_slot: FieldLike,
      note_type_id: FieldLike,
      serialized_note: FieldLike[],
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;
  };
}
