// @aztec/noir-contracts.js/src/SchnorrAccount.ts
/* Autogenerated file, do not edit! */

/* eslint-disable */
import {
  AztecAddress,
  AztecAddressLike,
  CompleteAddress,
  Contract,
  ContractArtifact,
  ContractBase,
  ContractFunctionInteraction,
  ContractInstanceWithAddress,
  ContractMethod,
  DeployMethod,
  EthAddress,
  EthAddressLike,
  FieldLike,
  Fr,
  FunctionSelectorLike,
  NoirCompiledContract,
  Point,
  PublicKey,
  Wallet,
  WrappedFieldLike,
  loadContractArtifact,
} from '@aztec/aztec.js';

import SchnorrAccountContractArtifactJson from '../artifacts/schnorr_account_contract-SchnorrAccount.json' assert { type: 'json' };

export const SchnorrAccountContractArtifact = loadContractArtifact(
  SchnorrAccountContractArtifactJson as NoirCompiledContract,
);

/**
 * Type-safe interface for contract SchnorrAccount;
 */
export class SchnorrAccountContract extends ContractBase {
  private constructor(instance: ContractInstanceWithAddress, wallet: Wallet) {
    super(instance, SchnorrAccountContractArtifact, wallet);
  }

  /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */
  public static async at(address: AztecAddress, wallet: Wallet) {
    return Contract.at(address, SchnorrAccountContract.artifact, wallet) as Promise<SchnorrAccountContract>;
  }

  /**
   * Creates a tx to deploy a new instance of this contract.
   */
  public static deploy(wallet: Wallet, signing_pub_key_x: FieldLike, signing_pub_key_y: FieldLike) {
    return new DeployMethod<SchnorrAccountContract>(
      Point.ZERO,
      wallet,
      SchnorrAccountContractArtifact,
      SchnorrAccountContract.at,
      Array.from(arguments).slice(1),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified public key to derive the address.
   */
  public static deployWithPublicKey(
    publicKey: PublicKey,
    wallet: Wallet,
    signing_pub_key_x: FieldLike,
    signing_pub_key_y: FieldLike,
  ) {
    return new DeployMethod<SchnorrAccountContract>(
      publicKey,
      wallet,
      SchnorrAccountContractArtifact,
      SchnorrAccountContract.at,
      Array.from(arguments).slice(2),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */
  public static deployWithOpts<M extends keyof SchnorrAccountContract['methods']>(
    opts: { publicKey?: PublicKey; method?: M; wallet: Wallet },
    ...args: Parameters<SchnorrAccountContract['methods'][M]>
  ) {
    return new DeployMethod<SchnorrAccountContract>(
      opts.publicKey ?? Point.ZERO,
      opts.wallet,
      SchnorrAccountContractArtifact,
      SchnorrAccountContract.at,
      Array.from(arguments).slice(1),
      opts.method ?? 'constructor',
    );
  }

  /**
   * Returns this contract's artifact.
   */
  public static get artifact(): ContractArtifact {
    return SchnorrAccountContractArtifact;
  }

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public methods!: {
    /** constructor(signing_pub_key_x: field, signing_pub_key_y: field) */
    constructor: ((signing_pub_key_x: FieldLike, signing_pub_key_y: FieldLike) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** entrypoint(app_payload: struct, fee_payload: struct) */
    entrypoint: ((
      app_payload: {
        function_calls: {
          args_hash: FieldLike;
          function_selector: FunctionSelectorLike;
          target_address: AztecAddressLike;
          is_public: boolean;
        }[];
        nonce: FieldLike;
      },
      fee_payload: {
        function_calls: {
          args_hash: FieldLike;
          function_selector: FunctionSelectorLike;
          target_address: AztecAddressLike;
          is_public: boolean;
        }[];
        nonce: FieldLike;
      },
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** spend_public_authwit(inner_hash: field) */
    spend_public_authwit: ((inner_hash: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** spend_private_authwit(inner_hash: field) */
    spend_private_authwit: ((inner_hash: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** compute_note_hash_and_nullifier(contract_address: struct, nonce: field, storage_slot: field, note_type_id: field, serialized_note: array) */
    compute_note_hash_and_nullifier: ((
      contract_address: AztecAddressLike,
      nonce: FieldLike,
      storage_slot: FieldLike,
      note_type_id: FieldLike,
      serialized_note: FieldLike[],
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;
  };
}
