// @aztec/aztec.js/dest/wallet/signerless_wallet.js
import { DefaultEntrypoint } from '../entrypoint/default_entrypoint.js';
import { BaseWallet } from './base_wallet.js';
/**
 * Wallet implementation which creates a transaction request directly to the requested contract without any signing.
 */
export class SignerlessWallet extends BaseWallet {
    constructor(pxe, entrypoint) {
        super(pxe);
        this.entrypoint = entrypoint;
    }
    async createTxExecutionRequest(executions) {
        let entrypoint = this.entrypoint;
        if (!entrypoint) {
            const { chainId, protocolVersion } = await this.pxe.getNodeInfo();
            entrypoint = new DefaultEntrypoint(chainId, protocolVersion);
        }
        return entrypoint.createTxExecutionRequest(executions);
    }
    getChainId() {
        throw new Error('Method not implemented.');
    }
    getVersion() {
        throw new Error('Method not implemented.');
    }
    getCompleteAddress() {
        throw new Error('Method not implemented.');
    }
    createAuthWit(_messageHash) {
        throw new Error('Method not implemented.');
    }
}
