// @aztec/noir-contracts.js/dest/src/Claim.js
/* Autogenerated file, do not edit! */
/* eslint-disable */
import { Contract, ContractBase, DeployMethod, Point, loadContractArtifact, } from '@aztec/aztec.js';
import ClaimContractArtifactJson from '../artifacts/claim_contract-Claim.json' assert { type: 'json' };
export const ClaimContractArtifact = loadContractArtifact(ClaimContractArtifactJson);
/**
 * Type-safe interface for contract Claim;
 */
export class ClaimContract extends ContractBase {
    constructor(instance, wallet) {
        super(instance, ClaimContractArtifact, wallet);
    }
    /**
     * Creates a contract instance.
     * @param address - The deployed contract's address.
     * @param wallet - The wallet to use when interacting with the contract.
     * @returns A promise that resolves to a new Contract instance.
     */
    static async at(address, wallet) {
        return Contract.at(address, ClaimContract.artifact, wallet);
    }
    /**
     * Creates a tx to deploy a new instance of this contract.
     */
    static deploy(wallet, target_contract, reward_token) {
        return new DeployMethod(Point.ZERO, wallet, ClaimContractArtifact, ClaimContract.at, Array.from(arguments).slice(1));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified public key to derive the address.
     */
    static deployWithPublicKey(publicKey, wallet, target_contract, reward_token) {
        return new DeployMethod(publicKey, wallet, ClaimContractArtifact, ClaimContract.at, Array.from(arguments).slice(2));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified constructor method.
     */
    static deployWithOpts(opts, ...args) {
        return new DeployMethod(opts.publicKey ?? Point.ZERO, opts.wallet, ClaimContractArtifact, ClaimContract.at, Array.from(arguments).slice(1), opts.method ?? 'constructor');
    }
    /**
     * Returns this contract's artifact.
     */
    static get artifact() {
        return ClaimContractArtifact;
    }
}
