// @aztec/noir-contracts.js/dest/src/DelegatedOn.js
/* Autogenerated file, do not edit! */
/* eslint-disable */
import { Contract, ContractBase, DeployMethod, Point, loadContractArtifact, } from '@aztec/aztec.js';
import DelegatedOnContractArtifactJson from '../artifacts/delegated_on_contract-DelegatedOn.json' assert { type: 'json' };
export const DelegatedOnContractArtifact = loadContractArtifact(DelegatedOnContractArtifactJson);
/**
 * Type-safe interface for contract DelegatedOn;
 */
export class DelegatedOnContract extends ContractBase {
    constructor(instance, wallet) {
        super(instance, DelegatedOnContractArtifact, wallet);
    }
    /**
     * Creates a contract instance.
     * @param address - The deployed contract's address.
     * @param wallet - The wallet to use when interacting with the contract.
     * @returns A promise that resolves to a new Contract instance.
     */
    static async at(address, wallet) {
        return Contract.at(address, DelegatedOnContract.artifact, wallet);
    }
    /**
     * Creates a tx to deploy a new instance of this contract.
     */
    static deploy(wallet) {
        return new DeployMethod(Point.ZERO, wallet, DelegatedOnContractArtifact, DelegatedOnContract.at, Array.from(arguments).slice(1));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified public key to derive the address.
     */
    static deployWithPublicKey(publicKey, wallet) {
        return new DeployMethod(publicKey, wallet, DelegatedOnContractArtifact, DelegatedOnContract.at, Array.from(arguments).slice(2));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified constructor method.
     */
    static deployWithOpts(opts, ...args) {
        return new DeployMethod(opts.publicKey ?? Point.ZERO, opts.wallet, DelegatedOnContractArtifact, DelegatedOnContract.at, Array.from(arguments).slice(1), opts.method ?? 'constructor');
    }
    /**
     * Returns this contract's artifact.
     */
    static get artifact() {
        return DelegatedOnContractArtifact;
    }
}
