// @aztec/noir-contracts.js/dest/src/EasyPrivateToken.js
/* Autogenerated file, do not edit! */
/* eslint-disable */
import { Contract, ContractBase, DeployMethod, Point, loadContractArtifact, } from '@aztec/aztec.js';
import EasyPrivateTokenContractArtifactJson from '../artifacts/easy_private_token_contract-EasyPrivateToken.json' assert { type: 'json' };
export const EasyPrivateTokenContractArtifact = loadContractArtifact(EasyPrivateTokenContractArtifactJson);
/**
 * Type-safe interface for contract EasyPrivateToken;
 */
export class EasyPrivateTokenContract extends ContractBase {
    constructor(instance, wallet) {
        super(instance, EasyPrivateTokenContractArtifact, wallet);
    }
    /**
     * Creates a contract instance.
     * @param address - The deployed contract's address.
     * @param wallet - The wallet to use when interacting with the contract.
     * @returns A promise that resolves to a new Contract instance.
     */
    static async at(address, wallet) {
        return Contract.at(address, EasyPrivateTokenContract.artifact, wallet);
    }
    /**
     * Creates a tx to deploy a new instance of this contract.
     */
    static deploy(wallet, initial_supply, owner) {
        return new DeployMethod(Point.ZERO, wallet, EasyPrivateTokenContractArtifact, EasyPrivateTokenContract.at, Array.from(arguments).slice(1));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified public key to derive the address.
     */
    static deployWithPublicKey(publicKey, wallet, initial_supply, owner) {
        return new DeployMethod(publicKey, wallet, EasyPrivateTokenContractArtifact, EasyPrivateTokenContract.at, Array.from(arguments).slice(2));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified constructor method.
     */
    static deployWithOpts(opts, ...args) {
        return new DeployMethod(opts.publicKey ?? Point.ZERO, opts.wallet, EasyPrivateTokenContractArtifact, EasyPrivateTokenContract.at, Array.from(arguments).slice(1), opts.method ?? 'constructor');
    }
    /**
     * Returns this contract's artifact.
     */
    static get artifact() {
        return EasyPrivateTokenContractArtifact;
    }
}
