// @aztec/noir-contracts.js/dest/src/EcdsaAccount.js
/* Autogenerated file, do not edit! */
/* eslint-disable */
import { Contract, ContractBase, DeployMethod, Point, loadContractArtifact, } from '@aztec/aztec.js';
import EcdsaAccountContractArtifactJson from '../artifacts/ecdsa_account_contract-EcdsaAccount.json' assert { type: 'json' };
export const EcdsaAccountContractArtifact = loadContractArtifact(EcdsaAccountContractArtifactJson);
/**
 * Type-safe interface for contract EcdsaAccount;
 */
export class EcdsaAccountContract extends ContractBase {
    constructor(instance, wallet) {
        super(instance, EcdsaAccountContractArtifact, wallet);
    }
    /**
     * Creates a contract instance.
     * @param address - The deployed contract's address.
     * @param wallet - The wallet to use when interacting with the contract.
     * @returns A promise that resolves to a new Contract instance.
     */
    static async at(address, wallet) {
        return Contract.at(address, EcdsaAccountContract.artifact, wallet);
    }
    /**
     * Creates a tx to deploy a new instance of this contract.
     */
    static deploy(wallet, signing_pub_key_x, signing_pub_key_y) {
        return new DeployMethod(Point.ZERO, wallet, EcdsaAccountContractArtifact, EcdsaAccountContract.at, Array.from(arguments).slice(1));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified public key to derive the address.
     */
    static deployWithPublicKey(publicKey, wallet, signing_pub_key_x, signing_pub_key_y) {
        return new DeployMethod(publicKey, wallet, EcdsaAccountContractArtifact, EcdsaAccountContract.at, Array.from(arguments).slice(2));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified constructor method.
     */
    static deployWithOpts(opts, ...args) {
        return new DeployMethod(opts.publicKey ?? Point.ZERO, opts.wallet, EcdsaAccountContractArtifact, EcdsaAccountContract.at, Array.from(arguments).slice(1), opts.method ?? 'constructor');
    }
    /**
     * Returns this contract's artifact.
     */
    static get artifact() {
        return EcdsaAccountContractArtifact;
    }
}
