// @aztec/noir-contracts.js/dest/src/FPC.js
/* Autogenerated file, do not edit! */
/* eslint-disable */
import { Contract, ContractBase, DeployMethod, Point, loadContractArtifact, } from '@aztec/aztec.js';
import FPCContractArtifactJson from '../artifacts/fpc_contract-FPC.json' assert { type: 'json' };
export const FPCContractArtifact = loadContractArtifact(FPCContractArtifactJson);
/**
 * Type-safe interface for contract FPC;
 */
export class FPCContract extends ContractBase {
    constructor(instance, wallet) {
        super(instance, FPCContractArtifact, wallet);
    }
    /**
     * Creates a contract instance.
     * @param address - The deployed contract's address.
     * @param wallet - The wallet to use when interacting with the contract.
     * @returns A promise that resolves to a new Contract instance.
     */
    static async at(address, wallet) {
        return Contract.at(address, FPCContract.artifact, wallet);
    }
    /**
     * Creates a tx to deploy a new instance of this contract.
     */
    static deploy(wallet, other_asset, gas_token_address) {
        return new DeployMethod(Point.ZERO, wallet, FPCContractArtifact, FPCContract.at, Array.from(arguments).slice(1));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified public key to derive the address.
     */
    static deployWithPublicKey(publicKey, wallet, other_asset, gas_token_address) {
        return new DeployMethod(publicKey, wallet, FPCContractArtifact, FPCContract.at, Array.from(arguments).slice(2));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified constructor method.
     */
    static deployWithOpts(opts, ...args) {
        return new DeployMethod(opts.publicKey ?? Point.ZERO, opts.wallet, FPCContractArtifact, FPCContract.at, Array.from(arguments).slice(1), opts.method ?? 'constructor');
    }
    /**
     * Returns this contract's artifact.
     */
    static get artifact() {
        return FPCContractArtifact;
    }
}
