// @aztec/noir-contracts.js/dest/src/PriceFeed.js
/* Autogenerated file, do not edit! */
/* eslint-disable */
import { Contract, ContractBase, DeployMethod, Point, loadContractArtifact, } from '@aztec/aztec.js';
import PriceFeedContractArtifactJson from '../artifacts/price_feed_contract-PriceFeed.json' assert { type: 'json' };
export const PriceFeedContractArtifact = loadContractArtifact(PriceFeedContractArtifactJson);
/**
 * Type-safe interface for contract PriceFeed;
 */
export class PriceFeedContract extends ContractBase {
    constructor(instance, wallet) {
        super(instance, PriceFeedContractArtifact, wallet);
    }
    /**
     * Creates a contract instance.
     * @param address - The deployed contract's address.
     * @param wallet - The wallet to use when interacting with the contract.
     * @returns A promise that resolves to a new Contract instance.
     */
    static async at(address, wallet) {
        return Contract.at(address, PriceFeedContract.artifact, wallet);
    }
    /**
     * Creates a tx to deploy a new instance of this contract.
     */
    static deploy(wallet) {
        return new DeployMethod(Point.ZERO, wallet, PriceFeedContractArtifact, PriceFeedContract.at, Array.from(arguments).slice(1));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified public key to derive the address.
     */
    static deployWithPublicKey(publicKey, wallet) {
        return new DeployMethod(publicKey, wallet, PriceFeedContractArtifact, PriceFeedContract.at, Array.from(arguments).slice(2));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified constructor method.
     */
    static deployWithOpts(opts, ...args) {
        return new DeployMethod(opts.publicKey ?? Point.ZERO, opts.wallet, PriceFeedContractArtifact, PriceFeedContract.at, Array.from(arguments).slice(1), opts.method ?? 'constructor');
    }
    /**
     * Returns this contract's artifact.
     */
    static get artifact() {
        return PriceFeedContractArtifact;
    }
}
