// @aztec/noir-contracts.js/dest/src/SchnorrSingleKeyAccount.js
/* Autogenerated file, do not edit! */
/* eslint-disable */
import { Contract, ContractBase, DeployMethod, Point, loadContractArtifact, } from '@aztec/aztec.js';
import SchnorrSingleKeyAccountContractArtifactJson from '../artifacts/schnorr_single_key_account_contract-SchnorrSingleKeyAccount.json' assert { type: 'json' };
export const SchnorrSingleKeyAccountContractArtifact = loadContractArtifact(SchnorrSingleKeyAccountContractArtifactJson);
/**
 * Type-safe interface for contract SchnorrSingleKeyAccount;
 */
export class SchnorrSingleKeyAccountContract extends ContractBase {
    constructor(instance, wallet) {
        super(instance, SchnorrSingleKeyAccountContractArtifact, wallet);
    }
    /**
     * Creates a contract instance.
     * @param address - The deployed contract's address.
     * @param wallet - The wallet to use when interacting with the contract.
     * @returns A promise that resolves to a new Contract instance.
     */
    static async at(address, wallet) {
        return Contract.at(address, SchnorrSingleKeyAccountContract.artifact, wallet);
    }
    /**
     * Creates a tx to deploy a new instance of this contract.
     */
    static deploy(wallet) {
        return new DeployMethod(Point.ZERO, wallet, SchnorrSingleKeyAccountContractArtifact, SchnorrSingleKeyAccountContract.at, Array.from(arguments).slice(1));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified public key to derive the address.
     */
    static deployWithPublicKey(publicKey, wallet) {
        return new DeployMethod(publicKey, wallet, SchnorrSingleKeyAccountContractArtifact, SchnorrSingleKeyAccountContract.at, Array.from(arguments).slice(2));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified constructor method.
     */
    static deployWithOpts(opts, ...args) {
        return new DeployMethod(opts.publicKey ?? Point.ZERO, opts.wallet, SchnorrSingleKeyAccountContractArtifact, SchnorrSingleKeyAccountContract.at, Array.from(arguments).slice(1), opts.method ?? 'constructor');
    }
    /**
     * Returns this contract's artifact.
     */
    static get artifact() {
        return SchnorrSingleKeyAccountContractArtifact;
    }
}
