// @aztec/noir-contracts.js/dest/src/TokenBlacklist.js
/* Autogenerated file, do not edit! */
/* eslint-disable */
import { Contract, ContractBase, DeployMethod, Point, loadContractArtifact, } from '@aztec/aztec.js';
import TokenBlacklistContractArtifactJson from '../artifacts/token_blacklist_contract-TokenBlacklist.json' assert { type: 'json' };
export const TokenBlacklistContractArtifact = loadContractArtifact(TokenBlacklistContractArtifactJson);
/**
 * Type-safe interface for contract TokenBlacklist;
 */
export class TokenBlacklistContract extends ContractBase {
    constructor(instance, wallet) {
        super(instance, TokenBlacklistContractArtifact, wallet);
    }
    /**
     * Creates a contract instance.
     * @param address - The deployed contract's address.
     * @param wallet - The wallet to use when interacting with the contract.
     * @returns A promise that resolves to a new Contract instance.
     */
    static async at(address, wallet) {
        return Contract.at(address, TokenBlacklistContract.artifact, wallet);
    }
    /**
     * Creates a tx to deploy a new instance of this contract.
     */
    static deploy(wallet, admin, slow_updates_contract) {
        return new DeployMethod(Point.ZERO, wallet, TokenBlacklistContractArtifact, TokenBlacklistContract.at, Array.from(arguments).slice(1));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified public key to derive the address.
     */
    static deployWithPublicKey(publicKey, wallet, admin, slow_updates_contract) {
        return new DeployMethod(publicKey, wallet, TokenBlacklistContractArtifact, TokenBlacklistContract.at, Array.from(arguments).slice(2));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified constructor method.
     */
    static deployWithOpts(opts, ...args) {
        return new DeployMethod(opts.publicKey ?? Point.ZERO, opts.wallet, TokenBlacklistContractArtifact, TokenBlacklistContract.at, Array.from(arguments).slice(1), opts.method ?? 'constructor');
    }
    /**
     * Returns this contract's artifact.
     */
    static get artifact() {
        return TokenBlacklistContractArtifact;
    }
}
