// @aztec/accounts/dest/defaults/account_interface.js
import { Fr } from '@aztec/circuits.js';
import { DefaultAccountEntrypoint } from '@aztec/entrypoints/account';
/**
 * Default implementation for an account interface. Requires that the account uses the default
 * entrypoint signature, which accept an AppPayload and a FeePayload as defined in noir-libs/aztec-noir/src/entrypoint module
 */
export class DefaultAccountInterface {
    constructor(authWitnessProvider, address, nodeInfo) {
        this.authWitnessProvider = authWitnessProvider;
        this.address = address;
        this.entrypoint = new DefaultAccountEntrypoint(address.address, authWitnessProvider, nodeInfo.chainId, nodeInfo.protocolVersion);
        this.chainId = new Fr(nodeInfo.chainId);
        this.version = new Fr(nodeInfo.protocolVersion);
    }
    createTxExecutionRequest(executions, fee) {
        return this.entrypoint.createTxExecutionRequest(executions, fee);
    }
    createAuthWit(messageHash) {
        return this.authWitnessProvider.createAuthWit(messageHash);
    }
    getCompleteAddress() {
        return this.address;
    }
    getAddress() {
        return this.address.address;
    }
    getChainId() {
        return this.chainId;
    }
    getVersion() {
        return this.version;
    }
}
