// @aztec/ethereum/dest/deploy_l1_contracts.js
import { EthAddress } from '@aztec/foundation/eth-address';
import { createPublicClient, createWalletClient, getAddress, getContract, http, } from 'viem';
/**
 * Deploys the aztec L1 contracts; Rollup, Contract Deployment Emitter & (optionally) Decoder Helper.
 * @param rpcUrl - URL of the ETH RPC to use for deployment.
 * @param account - Private Key or HD Account that will deploy the contracts.
 * @param chain - The chain instance to deploy to.
 * @param logger - A logger object.
 * @param contractsToDeploy - The set of L1 artifacts to be deployed
 * @returns A list of ETH addresses of the deployed contracts.
 */
export const deployL1Contracts = async (rpcUrl, account, chain, logger, contractsToDeploy) => {
    logger('Deploying contracts...');
    const walletClient = createWalletClient({
        account,
        chain,
        transport: http(rpcUrl),
    });
    const publicClient = createPublicClient({
        chain,
        transport: http(rpcUrl),
    });
    const registryAddress = await deployL1Contract(walletClient, publicClient, contractsToDeploy.registry.contractAbi, contractsToDeploy.registry.contractBytecode);
    logger(`Deployed Registry at ${registryAddress}`);
    const outboxAddress = await deployL1Contract(walletClient, publicClient, contractsToDeploy.outbox.contractAbi, contractsToDeploy.outbox.contractBytecode, [getAddress(registryAddress.toString())]);
    logger(`Deployed Outbox at ${outboxAddress}`);
    const availabilityOracleAddress = await deployL1Contract(walletClient, publicClient, contractsToDeploy.availabilityOracle.contractAbi, contractsToDeploy.availabilityOracle.contractBytecode);
    logger(`Deployed AvailabilityOracle at ${availabilityOracleAddress}`);
    const rollupAddress = await deployL1Contract(walletClient, publicClient, contractsToDeploy.rollup.contractAbi, contractsToDeploy.rollup.contractBytecode, [getAddress(registryAddress.toString()), getAddress(availabilityOracleAddress.toString())]);
    logger(`Deployed Rollup at ${rollupAddress}`);
    // Inbox is immutable and is deployed from Rollup's constructor so we just fetch it from the contract.
    let inboxAddress;
    {
        const rollup = getContract({
            address: getAddress(rollupAddress.toString()),
            abi: contractsToDeploy.rollup.contractAbi,
            client: publicClient,
        });
        inboxAddress = EthAddress.fromString((await rollup.read.INBOX([])));
    }
    // We need to call a function on the registry to set the various contract addresses.
    const registryContract = getContract({
        address: getAddress(registryAddress.toString()),
        abi: contractsToDeploy.registry.contractAbi,
        client: walletClient,
    });
    await registryContract.write.upgrade([getAddress(rollupAddress.toString()), getAddress(inboxAddress.toString()), getAddress(outboxAddress.toString())], { account });
    // this contract remains uninitialized because at this point we don't know the address of the gas token on L2
    const gasTokenAddress = await deployL1Contract(walletClient, publicClient, contractsToDeploy.gasToken.contractAbi, contractsToDeploy.gasToken.contractBytecode);
    logger(`Deployed Gas Token at ${gasTokenAddress}`);
    // this contract remains uninitialized because at this point we don't know the address of the gas token on L2
    const gasPortalAddress = await deployL1Contract(walletClient, publicClient, contractsToDeploy.gasPortal.contractAbi, contractsToDeploy.gasPortal.contractBytecode);
    logger(`Deployed Gas Portal at ${gasPortalAddress}`);
    const l1Contracts = {
        availabilityOracleAddress,
        rollupAddress,
        registryAddress,
        inboxAddress,
        outboxAddress,
        gasTokenAddress,
        gasPortalAddress,
    };
    return {
        walletClient,
        publicClient,
        l1ContractAddresses: l1Contracts,
    };
};
// docs:start:deployL1Contract
/**
 * Helper function to deploy ETH contracts.
 * @param walletClient - A viem WalletClient.
 * @param publicClient - A viem PublicClient.
 * @param abi - The ETH contract's ABI (as abitype's Abi).
 * @param bytecode  - The ETH contract's bytecode.
 * @param args - Constructor arguments for the contract.
 * @returns The ETH address the contract was deployed to.
 */
export async function deployL1Contract(walletClient, publicClient, abi, bytecode, args = []) {
    const hash = await walletClient.deployContract({
        abi,
        bytecode,
        args,
    });
    const receipt = await publicClient.waitForTransactionReceipt({ hash });
    const contractAddress = receipt.contractAddress;
    if (!contractAddress) {
        throw new Error(`No contract address found in receipt: ${JSON.stringify(receipt)}`);
    }
    return EthAddress.fromString(receipt.contractAddress);
}
// docs:end:deployL1Contract
