// @aztec/circuit-types/dest/interfaces/nullifier_tree.js
import { Fr, NullifierLeafPreimage } from '@aztec/circuits.js';
import { SiblingPath } from '../sibling_path/index.js';
/**
 * Nullifier membership witness.
 * @remarks When this represents membership witness of a low nullifier it can be used to perform a nullifier
 * non-inclusion proof by leveraging the "linked list structure" of leaves and proving that a lower nullifier
 * is pointing to a bigger next value than the nullifier we are trying to prove non-inclusion for.
 */
export class NullifierMembershipWitness {
    constructor(
    /**
     * The index of the nullifier in a tree.
     */
    index, 
    /**
     * Preimage of the nullifier.
     */
    leafPreimage, 
    /**
     * Sibling path to prove membership of the nullifier.
     */
    siblingPath) {
        this.index = index;
        this.leafPreimage = leafPreimage;
        this.siblingPath = siblingPath;
    }
    /**
     * Returns a field array representation of a nullifier witness.
     * @returns A field array representation of a nullifier witness.
     */
    toFields() {
        return [new Fr(this.index), ...this.leafPreimage.toFields(), ...this.siblingPath.toFields()];
    }
    toJSON() {
        return {
            index: '0x' + this.index.toString(16),
            leafPreimage: this.leafPreimage.toJSON(),
            siblingPath: this.siblingPath.toString(),
        };
    }
    static fromJSON(json) {
        return new NullifierMembershipWitness(BigInt(json.index), NullifierLeafPreimage.fromJSON(json.leafPreimage), SiblingPath.fromString(json.siblingPath));
    }
}
